{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "import re\n",
    "import sys\n",
    "import h5py\n",
    "from math import ceil\n",
    "\n",
    "CL_max = 100000\n",
    "SL=5000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyfastx\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gffutils.create_db('/odinn/tmp/benediktj/Data/SplicePrediction/gencode.v39.annotation.gtf', \"/odinn/tmp/benediktj/Data/SplicePrediction/gencode.v39.annotation.db\", force=True,disable_infer_genes=True, disable_infer_transcripts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gtf = gffutils.FeatureDB(\"/odinn/tmp/benediktj/Data/SplicePrediction/gencode.v39.annotation.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fasta = pyfastx.Fasta('/odinn/tmp/benediktj/SpliceAITrainingCode/hg38.fa')\n",
    "fasta = pyfastx.Fasta('/odinn/tmp/benediktj/SpliceAITrainingCode/hg19.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_test = pd.read_csv('/odinn/tmp/benediktj/SpliceAITrainingCode/gencode_test.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#canonical_data.columns = ['name','?','chrm','strand','tx_start','tx_end','acceptor','donor']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>exonStarts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AZIN2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>33546714</td>\n",
       "      <td>33585995</td>\n",
       "      <td>33546895,33547109,33547413,33547955,33549728,3...</td>\n",
       "      <td>33546989,33547202,33547779,33549555,33557651,3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRUNE</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>150980953</td>\n",
       "      <td>151008189</td>\n",
       "      <td>150981147,150990380,150991145,150997271,150998...</td>\n",
       "      <td>150990288,150990943,150997087,150997991,150999...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C1orf21</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>184356192</td>\n",
       "      <td>184598154</td>\n",
       "      <td>184356502,184446737,184476816,184559949,184567...</td>\n",
       "      <td>184446520,184476722,184559873,184567535,184588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LIN9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>226418858</td>\n",
       "      <td>226497434</td>\n",
       "      <td>226420307,226420894,226421224,226426797,226438...</td>\n",
       "      <td>226420797,226421045,226426672,226438540,226453...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf159</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1017203</td>\n",
       "      <td>1051478</td>\n",
       "      <td>1018367,1019763,1019886,1021392,1022584,102297...</td>\n",
       "      <td>1019733,1019861,1021258,1022519,1022882,102573...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1647</th>\n",
       "      <td>NELFB</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>140149759</td>\n",
       "      <td>140167999</td>\n",
       "      <td>140150063,140150522,140150880,140151506,140157...</td>\n",
       "      <td>140150359,140150781,140151276,140157489,140158...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648</th>\n",
       "      <td>NSMF</td>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>140342025</td>\n",
       "      <td>140353786</td>\n",
       "      <td>140343943,140344126,140344477,140344707,140346...</td>\n",
       "      <td>140344051,140344375,140344628,140346817,140347...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1649</th>\n",
       "      <td>MRPL41</td>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>140445651</td>\n",
       "      <td>140447007</td>\n",
       "      <td>140446430,</td>\n",
       "      <td>140446526,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650</th>\n",
       "      <td>DPH7</td>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>140449361</td>\n",
       "      <td>140473387</td>\n",
       "      <td>140450100,140459058,140459410,140459606,140468...</td>\n",
       "      <td>140458886,140459345,140459537,140468660,140469...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651</th>\n",
       "      <td>ZMYND19</td>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>140476531</td>\n",
       "      <td>140484942</td>\n",
       "      <td>140477138,140477615,140481559,140482275,140483...</td>\n",
       "      <td>140477435,140481419,140482169,140483111,140484...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1652 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #name chrom strand    txStart      txEnd  \\\n",
       "0        AZIN2  chr1      +   33546714   33585995   \n",
       "1        PRUNE  chr1      +  150980953  151008189   \n",
       "2      C1orf21  chr1      +  184356192  184598154   \n",
       "3         LIN9  chr1      -  226418858  226497434   \n",
       "4     C1orf159  chr1      -    1017203    1051478   \n",
       "...        ...   ...    ...        ...        ...   \n",
       "1647     NELFB  chr9      +  140149759  140167999   \n",
       "1648      NSMF  chr9      -  140342025  140353786   \n",
       "1649    MRPL41  chr9      +  140445651  140447007   \n",
       "1650      DPH7  chr9      -  140449361  140473387   \n",
       "1651   ZMYND19  chr9      -  140476531  140484942   \n",
       "\n",
       "                                               exonEnds  \\\n",
       "0     33546895,33547109,33547413,33547955,33549728,3...   \n",
       "1     150981147,150990380,150991145,150997271,150998...   \n",
       "2     184356502,184446737,184476816,184559949,184567...   \n",
       "3     226420307,226420894,226421224,226426797,226438...   \n",
       "4     1018367,1019763,1019886,1021392,1022584,102297...   \n",
       "...                                                 ...   \n",
       "1647  140150063,140150522,140150880,140151506,140157...   \n",
       "1648  140343943,140344126,140344477,140344707,140346...   \n",
       "1649                                         140446430,   \n",
       "1650  140450100,140459058,140459410,140459606,140468...   \n",
       "1651  140477138,140477615,140481559,140482275,140483...   \n",
       "\n",
       "                                             exonStarts  \n",
       "0     33546989,33547202,33547779,33549555,33557651,3...  \n",
       "1     150990288,150990943,150997087,150997991,150999...  \n",
       "2     184446520,184476722,184559873,184567535,184588...  \n",
       "3     226420797,226421045,226426672,226438540,226453...  \n",
       "4     1019733,1019861,1021258,1022519,1022882,102573...  \n",
       "...                                                 ...  \n",
       "1647  140150359,140150781,140151276,140157489,140158...  \n",
       "1648  140344051,140344375,140344628,140346817,140347...  \n",
       "1649                                         140446526,  \n",
       "1650  140458886,140459345,140459537,140468660,140469...  \n",
       "1651  140477435,140481419,140482169,140483111,140484...  \n",
       "\n",
       "[1652 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_datapoints(seq, strand, tx_start, tx_end, jn_start, jn_end):\n",
    "    # This function first converts the sequence into an integer array, where\n",
    "    # A, C, G, T, N are mapped to 1, 2, 3, 4, 0 respectively. If the strand is\n",
    "    # negative, then reverse complementing is done. The splice junctions \n",
    "    # are also converted into an array of integers, where 0, 1, 2, -1 \n",
    "    # correspond to no splicing, acceptor, donor and missing information\n",
    "    # respectively. It then calls reformat_data and one_hot_encode\n",
    "    # and returns X, Y which can be used by Keras models.\n",
    "    #seq = fasta[chrom][int(gene_start)-CL_max//2-1:int(gene_end)+CL_max//2]\n",
    "    #seq = seq.seq\n",
    "    #seq = 'N'*(CL_max//2) + seq[CL_max//2:-CL_max//2] + 'N'*(CL_max//2)\n",
    "    seq = 'N'*(CL_max//2) + seq + 'N'*(CL_max//2)\n",
    "    # Context being provided on the RNA and not the DNA\n",
    "    #.replace('R', '0').replace('M', '0').replace('B', '0').replace('W', '0')\n",
    "    seq = seq.upper()\n",
    "    seq = re.sub(r'[^AGTC]', '0',seq)\n",
    "    seq = seq.replace('A', '1').replace('C', '2')\n",
    "    seq = seq.replace('G', '3').replace('T', '4').replace('N', '0')\n",
    "\n",
    "\n",
    "    tx_start = int(tx_start)\n",
    "    tx_end = int(tx_end) \n",
    "\n",
    "    jn_start = [[x for x in jn_start]]\n",
    "    jn_end = [[x for x in jn_end]]\n",
    "    \n",
    "    if strand == '+':\n",
    "\n",
    "        X0 = np.asarray([int(x) for x in seq])\n",
    "        Y0 = [-np.ones(tx_end-tx_start+1)]\n",
    "        \n",
    "        for t in range(1):\n",
    "            \n",
    "            if len(jn_start[t]) > 0:\n",
    "                Y0[t] = np.zeros(tx_end-tx_start+1)\n",
    "                for c in jn_start[t]:\n",
    "                    if tx_start <= c <= tx_end:\n",
    "                        Y0[t][c-tx_start] = 2\n",
    "                for c in jn_end[t]:\n",
    "                    if tx_start <= c <= tx_end:\n",
    "                        Y0[t][c-tx_start] = 1\n",
    "                    # Ignoring junctions outside annotated tx start/end sites\n",
    "                     \n",
    "    elif strand == '-':\n",
    "\n",
    "        X0 = (5-np.asarray([int(x) for x in seq[::-1]])) % 5  # Reverse complement\n",
    "        Y0 = [-np.ones(tx_end-tx_start+1)]\n",
    "\n",
    "        for t in range(1):\n",
    "\n",
    "            if len(jn_start[t]) > 0:\n",
    "                Y0[t] = np.zeros(tx_end-tx_start+1)\n",
    "                for c in jn_end[t]:\n",
    "                    if tx_start <= c <= tx_end:\n",
    "                        Y0[t][tx_end-c] = 2\n",
    "                for c in jn_start[t]:\n",
    "                    if tx_start <= c <= tx_end:\n",
    "                        Y0[t][tx_end-c] = 1\n",
    "    \n",
    "\n",
    "    Xd, Yd = reformat_data(X0, Y0)\n",
    "    X, Y = one_hot_encode(Xd, Yd)\n",
    "\n",
    "    return X, Y\n",
    "\n",
    "def reformat_data(X0, Y0):\n",
    "    # This function converts X0, Y0 of the create_datapoints function into\n",
    "    # blocks such that the data is broken down into data points where the\n",
    "    # input is a sequence of length SL+CL_max corresponding to SL nucleotides\n",
    "    # of interest and CL_max context nucleotides, the output is a sequence of\n",
    "    # length SL corresponding to the splicing information of the nucleotides\n",
    "    # of interest. The CL_max context nucleotides are such that they are\n",
    "    # CL_max/2 on either side of the SL nucleotides of interest.\n",
    "\n",
    "    num_points = ceil_div(len(Y0[0]), SL)\n",
    "\n",
    "    Xd = np.zeros((num_points, SL+CL_max))\n",
    "    Yd = [-np.ones((num_points, SL)) for t in range(1)]\n",
    "\n",
    "    X0 = np.pad(X0, [0, SL], 'constant', constant_values=0)\n",
    "    Y0 = [np.pad(Y0[t], [0, SL], 'constant', constant_values=-1) for t in range(1)]\n",
    "\n",
    "    for i in range(num_points):\n",
    "        Xd[i] = X0[SL*i:CL_max+SL*(i+1)]\n",
    "\n",
    "    for i in range(num_points):\n",
    "        Yd[0][i] = Y0[0][SL*i:SL*(i+1)]\n",
    "\n",
    "    return Xd, Yd\n",
    "\n",
    "def ceil_div(x, y):\n",
    "\n",
    "    return int(ceil(float(x)/y))\n",
    "\n",
    "\n",
    "IN_MAP = np.asarray([[0, 0, 0, 0],\n",
    "                     [1, 0, 0, 0],\n",
    "                     [0, 1, 0, 0],\n",
    "                     [0, 0, 1, 0],\n",
    "                     [0, 0, 0, 1]])\n",
    "# One-hot encoding of the inputs: 0 is for padding, and 1, 2, 3, 4 correspond\n",
    "# to A, C, G, T respectively.\n",
    "\n",
    "OUT_MAP = np.asarray([[1, 0, 0],\n",
    "                      [0, 1, 0],\n",
    "                      [0, 0, 1],\n",
    "                      [0, 0, 0]])\n",
    "\n",
    "def one_hot_encode(Xd, Yd):\n",
    "\n",
    "    return IN_MAP[Xd.astype('int8')], \\\n",
    "           [OUT_MAP[Yd[t].astype('int8')] for t in range(1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def createDataset(gencode,setType,data_dir):\n",
    "    h5f2 = h5py.File(data_dir + 'gencode_{}k_dataset'.format(CL_max//1000)\n",
    "                        + '_' + setType + '_' \n",
    "                        + '.h5', 'w')\n",
    "\n",
    "    if setType == 'train':\n",
    "        CHROM_GROUP = ['chr11', 'chr13', 'chr15', 'chr17', 'chr19', 'chr21',\n",
    "                           'chr2', 'chr4', 'chr6', 'chr8', 'chr10', 'chr12',\n",
    "                           'chr14', 'chr16', 'chr18', 'chr20', 'chr22', 'chrX', 'chrY']\n",
    "    elif setType == 'test':\n",
    "        CHROM_GROUP = ['chr1', 'chr3', 'chr5', 'chr7', 'chr9']\n",
    "    else:\n",
    "        CHROM_GROUP = ['chr1', 'chr3', 'chr5', 'chr7', 'chr9',\n",
    "                       'chr11', 'chr13', 'chr15', 'chr17', 'chr19', 'chr21',\n",
    "                       'chr2', 'chr4', 'chr6', 'chr8', 'chr10', 'chr12',\n",
    "                       'chr14', 'chr16', 'chr18', 'chr20', 'chr22', 'chrX', 'chrY']\n",
    "\n",
    "    CHUNK_SIZE = 100\n",
    "    idx = 0\n",
    "    \n",
    "    #if os.path.exists('{}/annotation_gencode_v39_{}.txt'.format(data_dir,setType)):\n",
    "    #    os.remove('{}/annotation_gencode_v39_{}.txt'.format(data_dir,setType))\n",
    "\n",
    "    for gene_nr in tqdm(range(gencode.shape[0])):\n",
    "        chrom = gencode.iloc[gene_nr,1]\n",
    "        strand = gencode.iloc[gene_nr,2]\n",
    "        #print(gene)\n",
    "        exon_junctions = []\n",
    "        tx_start = int(gencode.iloc[gene_nr,3])\n",
    "        tx_end = int(gencode.iloc[gene_nr,4])\n",
    "        jn_start = [int(x) for x in gencode.iloc[gene_nr,5].split(',')[:-1]]\n",
    "        jn_end = [int(x) for x in gencode.iloc[gene_nr,6].split(',')[:-1]]\n",
    "        \n",
    "\n",
    "        try:\n",
    "            seq = fasta[chrom][int(tx_start)-1:int(tx_end)]\n",
    "            seq = seq.seq\n",
    "        except:\n",
    "            print('Failed reading fasta file for {}:{}-{}'.format(chrom,tx_start,tx_end))\n",
    "            print('SKIPPING')\n",
    "            continue\n",
    "            \n",
    "            \n",
    "        X, Y = create_datapoints(seq, strand,tx_start,tx_end, jn_start, jn_end)\n",
    "        if idx%CHUNK_SIZE==0:\n",
    "            X_batch = []\n",
    "            Y_batch = [[]]\n",
    "\n",
    "        X_batch.extend(X)\n",
    "        Y_batch[0].extend(Y[0])\n",
    "\n",
    "        if idx%CHUNK_SIZE==CHUNK_SIZE-1:\n",
    "            X_batch = np.asarray(X_batch).astype('int8')\n",
    "            Y_batch[0] = np.asarray(Y_batch[0]).astype('int8')\n",
    "            lastIdxSave = idx\n",
    "            h5f2.create_dataset('X' + str(idx//CHUNK_SIZE), data=X_batch)\n",
    "            h5f2.create_dataset('Y' + str(idx//CHUNK_SIZE), data=Y_batch)\n",
    "        idx += 1\n",
    "\n",
    "    if lastIdxSave != idx-1:\n",
    "        X_batch = np.asarray(X_batch).astype('int8')\n",
    "        Y_batch[0] = np.asarray(Y_batch[0]).astype('int8')\n",
    "        h5f2.create_dataset('X' + str(lastIdxSave//CHUNK_SIZE+1), data=X_batch)\n",
    "        h5f2.create_dataset('Y' + str(lastIdxSave//CHUNK_SIZE+1), data=Y_batch)\n",
    "\n",
    "    h5f2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/odinn/tmp/benediktj/Data/SplicePrediction-050422/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1652/1652 [04:09<00:00,  6.63it/s]\n"
     ]
    }
   ],
   "source": [
    "createDataset(gencode_test,'test',data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "58051it [11:50, 81.74it/s]  \n"
     ]
    }
   ],
   "source": [
    "#createDataset('train',data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
