{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1c3bae6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benediktj/Python3.9.1/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import gmean\n",
    "from tqdm import tqdm\n",
    "import gffutils\n",
    "import os.path\n",
    "#from liftover import ChainFile\n",
    "from collections import defaultdict\n",
    "from math import ceil\n",
    "import pyfastx\n",
    "import re\n",
    "from scipy.sparse.csgraph import connected_components\n",
    "from glob import glob\n",
    "from scipy.sparse import lil_matrix,csr_matrix,coo_matrix,dok_matrix, save_npz\n",
    "import pickle\n",
    "#\n",
    "blacklist = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/ENCFF220FIN.bed',header=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cab6af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "simpileRepeats = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/simpleRepeat.txt',header=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "88746baa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gffutils.create_db('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX/gencode.v42.annotation.gtf', \"/odinn/tmp/benediktj/Data/SplicePrediction-GTEX/gencode.v42.annotation.db\", force=True,disable_infer_genes=True, disable_infer_transcripts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d829cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gtf_gencode = gffutils.FeatureDB('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/gencode.v42.annotation.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e71ed15",
   "metadata": {},
   "outputs": [],
   "source": [
    "leafcutter_counts = glob('/nfs/odinn/data/dataprocessing/rnasplice-blood/curry/leafCutter/*counts.tab')\n",
    "leafcutter_summary = glob('/nfs/odinn/data/dataprocessing/rnasplice-blood/curry/leafCutter/*summary.tab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9678bea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.read_csv('/odinn/users/gislih/RNA/requests/rna_paper/splice_anno/junction_db.tsv.gz', compression='gzip', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b5adf3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#junctions = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions.gct', nrows=10, skiprows=2,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "07d62be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#junctions = junctions.sort_values('Chr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e64dad7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findLowCountJunctions(leafcutter_counts,blacklist,simpileRepeats):\n",
    "    discardJunctionDict = defaultdict(bool)\n",
    "    discardReason = defaultdict(bool)\n",
    "    for f_i,file in enumerate(leafcutter_counts):\n",
    "        junctions = pd.read_csv(file,sep='\\t')\n",
    "        chrom =leafcutter_counts[f_i].split('_')[1]\n",
    "        print(chrom)\n",
    "        tmp2 = junctions.iloc[:,1:]\n",
    "        transcriptCount = np.sum(tmp2,axis=1)\n",
    "        tmp2.columns = [x.split('-')[1] for x in junctions.columns[1:]]\n",
    "        tmp2 = tmp2.T.groupby(tmp2.columns).sum().T\n",
    "\n",
    "        includeJunction = np.sum(tmp2 > 0,axis=1)>=4\n",
    "\n",
    "        blacklist_chrom = blacklist[blacklist[0]==chrom]\n",
    "        simpileRepeats_chrom = simpileRepeats[simpileRepeats[1]==chrom]\n",
    "\n",
    "        for i,junction in tqdm(enumerate(junctions.id.values)):\n",
    "            chrom,start,end,cluster = junctions.iloc[i,0].split(':')\n",
    "            start,end = int(start),int(end)\n",
    "            cond1 = includeJunction[i]==False\n",
    "\n",
    "            if cond1:\n",
    "                discardJunctionDict[junction] = True\n",
    "                discardReason[junction] = 'LowReadCount'\n",
    "                #prev_chrom = chrom\n",
    "                continue\n",
    "\n",
    "            #if prev_chrom != chrom:\n",
    "             #   blacklist_chrom = blacklist[blacklist[0]==chrom]\n",
    "            tmp = blacklist_chrom[np.logical_not(np.any([blacklist_chrom[1]>=end, blacklist_chrom[2]<=start],0))]\n",
    "            cond2 = False\n",
    "            for i_b in range(tmp.shape[0]):\n",
    "                if (np.isin(start, range(tmp.iloc[i_b,1],tmp.iloc[i_b,2]+1)) or np.isin(end, range(tmp.iloc[i_b,1],tmp.iloc[i_b,2]+1))):\n",
    "                    cond2 = True\n",
    "                    break\n",
    "\n",
    "            if cond2:\n",
    "                discardJunctionDict[junction] = True\n",
    "                discardReason[junction] = 'InBlacklistedRegion'\n",
    "                #prev_chrom = chrom\n",
    "                continue\n",
    "\n",
    "            #if prev_chrom != chrom:\n",
    "            #    simpileRepeats_chrom = simpileRepeats[simpileRepeats[1]==chrom]\n",
    "\n",
    "            tmp = simpileRepeats_chrom[np.logical_not(np.any([simpileRepeats_chrom[2]>=end,simpileRepeats_chrom[3]<=start],0))]\n",
    "            cond3 = False\n",
    "            for i_b in range(tmp.shape[0]):\n",
    "                if (np.isin(start, range(tmp.iloc[i_b,2],tmp.iloc[i_b,3]+1)) or np.isin(end, range(tmp.iloc[i_b,2],tmp.iloc[i_b,3]+1))):\n",
    "                    cond3 = True\n",
    "                    break\n",
    "\n",
    "            if cond3:\n",
    "                discardJunctionDict[junction] = True\n",
    "                discardReason[junction] = 'InRepeatRegion'\n",
    "                continue\n",
    "        \n",
    "    return discardJunctionDict,discardReason"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "98d2fb56",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10380it [00:14, 709.97it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "15511it [00:17, 901.36it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14781it [00:17, 835.26it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "4328it [00:05, 736.30it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "9351it [00:11, 838.48it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "8889it [00:10, 842.78it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13930it [00:16, 853.87it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "17646it [00:20, 864.72it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3751it [00:05, 730.32it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "18765it [00:20, 903.87it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25877it [00:32, 799.69it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7311it [00:08, 854.99it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3348it [00:04, 718.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7807it [00:08, 896.95it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "18279it [00:24, 752.57it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "15822it [00:21, 749.44it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "9260it [00:12, 739.58it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10859it [00:14, 758.79it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12945it [00:16, 771.95it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12764it [00:18, 698.28it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "8570it [00:12, 694.76it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10551it [00:13, 791.41it/s] \n"
     ]
    }
   ],
   "source": [
    "discardJunctionDict,discardReason = findLowCountJunctions(leafcutter_counts,blacklist,simpileRepeats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ac6fdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/discardJunctions.pkl', 'rb') as f:\n",
    "    discardJunctionDict = pickle.load(f)\n",
    "\n",
    "with open('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/discardReason.pkl', 'rb') as f:\n",
    "     discardJunctionDict = pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fe5b84df",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/discardJunctions.pkl', 'wb') as f:\n",
    "#    pickle.dump(discardJunctionDict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f1d37a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/discardReason.pkl', 'wb') as f:\n",
    "#    pickle.dump(discardReason, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a411b6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#leafcutterFiles = glob('/odinn/tmp/bjarnih/RNA/leafCutter/GTEx/*/meta_results/clusters_*_summary.tab')\n",
    "leafcutterFiles = leafcutter_summary\n",
    "chrmToLeafcutterFiles = defaultdict(list)\n",
    "for file in leafcutterFiles:\n",
    "    chrm = file.split('/')[-1].split('_')[1]\n",
    "    chrmToLeafcutterFiles[chrm].append(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "edc6fbf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCombinedLeafCutterDF(chrm,chrmToLeafcutterFiles):\n",
    "    files = chrmToLeafcutterFiles[chrm]\n",
    "    for i in range(len(files)):\n",
    "        if i == 0:\n",
    "            df = pd.read_csv(files[i],sep='\\t')\n",
    "        else:\n",
    "            df = pd.concat([df,pd.read_csv(files[i],sep='\\t')],axis=0)\n",
    "            df = df.drop_duplicates('splice_event_id')\n",
    "    return df.sort_values('Start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e2ce30ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = getCombinedLeafCutterDF('chr1',chrmToLeafcutterFiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c24a40b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ClusterID</th>\n",
       "      <th>splice_event_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_start_min</th>\n",
       "      <th>Gene_end_max</th>\n",
       "      <th>...</th>\n",
       "      <th>cluster_type</th>\n",
       "      <th>median_psi</th>\n",
       "      <th>max_psi</th>\n",
       "      <th>prop_NA</th>\n",
       "      <th>prop_zero</th>\n",
       "      <th>num_unique</th>\n",
       "      <th>low_variability</th>\n",
       "      <th>any_annotated_in_cluster</th>\n",
       "      <th>any_gene_expressed</th>\n",
       "      <th>protein_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183015</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:182746:183015:0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>+</td>\n",
       "      <td>182393</td>\n",
       "      <td>184158</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.467503</td>\n",
       "      <td>0.730008</td>\n",
       "      <td>271</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183113</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:182746:183113:0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>+</td>\n",
       "      <td>182393</td>\n",
       "      <td>184158</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.467503</td>\n",
       "      <td>0.191604</td>\n",
       "      <td>521</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183131</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:182746:183131:0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>+</td>\n",
       "      <td>182393</td>\n",
       "      <td>184158</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.467503</td>\n",
       "      <td>0.000631</td>\n",
       "      <td>536</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183164</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:182746:183164:0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>+</td>\n",
       "      <td>182393</td>\n",
       "      <td>184158</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.467503</td>\n",
       "      <td>0.825021</td>\n",
       "      <td>231</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>749381</td>\n",
       "      <td>752447</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:749381:752447:1</td>\n",
       "      <td>RP11-206L10.2,RP5-857K21.4</td>\n",
       "      <td>ENSG00000228327,ENSG00000230021</td>\n",
       "      <td>-</td>\n",
       "      <td>585989</td>\n",
       "      <td>827796</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_donor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.816450</td>\n",
       "      <td>0.578144</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25872</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248906342</td>\n",
       "      <td>248913815</td>\n",
       "      <td>4935</td>\n",
       "      <td>chr1:248906342:248913815:4935</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>+</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>...</td>\n",
       "      <td>alt_donor</td>\n",
       "      <td>0.971429</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010982</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>600</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25874</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248913879</td>\n",
       "      <td>248916601</td>\n",
       "      <td>4936</td>\n",
       "      <td>chr1:248913879:248916601:4936</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>+</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>...</td>\n",
       "      <td>complex</td>\n",
       "      <td>0.474788</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.020731</td>\n",
       "      <td>0.006923</td>\n",
       "      <td>1257</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25875</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248913879</td>\n",
       "      <td>248917278</td>\n",
       "      <td>4936</td>\n",
       "      <td>chr1:248913879:248917278:4936</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>+</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>...</td>\n",
       "      <td>complex</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.020731</td>\n",
       "      <td>0.002575</td>\n",
       "      <td>1262</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25873</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248913879</td>\n",
       "      <td>248914438</td>\n",
       "      <td>4936</td>\n",
       "      <td>chr1:248913879:248914438:4936</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>+</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>...</td>\n",
       "      <td>complex</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.020731</td>\n",
       "      <td>0.349754</td>\n",
       "      <td>573</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25876</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248914559</td>\n",
       "      <td>248916601</td>\n",
       "      <td>4936</td>\n",
       "      <td>chr1:248914559:248916601:4936</td>\n",
       "      <td>PGBD2</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>+</td>\n",
       "      <td>248906196</td>\n",
       "      <td>248919946</td>\n",
       "      <td>...</td>\n",
       "      <td>complex</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.020731</td>\n",
       "      <td>0.389690</td>\n",
       "      <td>550</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25877 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom      Start        End  ClusterID                splice_event_id  \\\n",
       "0      chr1     182746     183015          0           chr1:182746:183015:0   \n",
       "1      chr1     182746     183113          0           chr1:182746:183113:0   \n",
       "2      chr1     182746     183131          0           chr1:182746:183131:0   \n",
       "3      chr1     182746     183164          0           chr1:182746:183164:0   \n",
       "4      chr1     749381     752447          1           chr1:749381:752447:1   \n",
       "...     ...        ...        ...        ...                            ...   \n",
       "25872  chr1  248906342  248913815       4935  chr1:248906342:248913815:4935   \n",
       "25874  chr1  248913879  248916601       4936  chr1:248913879:248916601:4936   \n",
       "25875  chr1  248913879  248917278       4936  chr1:248913879:248917278:4936   \n",
       "25873  chr1  248913879  248914438       4936  chr1:248913879:248914438:4936   \n",
       "25876  chr1  248914559  248916601       4936  chr1:248914559:248916601:4936   \n",
       "\n",
       "                        Gene_name                          Gene_id Strand  \\\n",
       "0                      FO538757.2                  ENSG00000279928      +   \n",
       "1                      FO538757.2                  ENSG00000279928      +   \n",
       "2                      FO538757.2                  ENSG00000279928      +   \n",
       "3                      FO538757.2                  ENSG00000279928      +   \n",
       "4      RP11-206L10.2,RP5-857K21.4  ENSG00000228327,ENSG00000230021      -   \n",
       "...                           ...                              ...    ...   \n",
       "25872                       PGBD2                  ENSG00000185220      +   \n",
       "25874                       PGBD2                  ENSG00000185220      +   \n",
       "25875                       PGBD2                  ENSG00000185220      +   \n",
       "25873                       PGBD2                  ENSG00000185220      +   \n",
       "25876                       PGBD2                  ENSG00000185220      +   \n",
       "\n",
       "       Gene_start_min  Gene_end_max  ...  cluster_type median_psi   max_psi  \\\n",
       "0              182393        184158  ...  alt_acceptor   0.000000  1.000000   \n",
       "1              182393        184158  ...  alt_acceptor   0.166667  1.000000   \n",
       "2              182393        184158  ...  alt_acceptor   0.800000  1.000000   \n",
       "3              182393        184158  ...  alt_acceptor   0.000000  0.714286   \n",
       "4              585989        827796  ...     alt_donor   0.000000  1.000000   \n",
       "...               ...           ...  ...           ...        ...       ...   \n",
       "25872       248906196     248919946  ...     alt_donor   0.971429  1.000000   \n",
       "25874       248906196     248919946  ...       complex   0.474788  1.000000   \n",
       "25875       248906196     248919946  ...       complex   0.428571  1.000000   \n",
       "25873       248906196     248919946  ...       complex   0.040000  0.750000   \n",
       "25876       248906196     248919946  ...       complex   0.033333  0.600000   \n",
       "\n",
       "        prop_NA  prop_zero  num_unique  low_variability  \\\n",
       "0      0.467503   0.730008         271            False   \n",
       "1      0.467503   0.191604         521            False   \n",
       "2      0.467503   0.000631         536            False   \n",
       "3      0.467503   0.825021         231            False   \n",
       "4      0.816450   0.578144          75            False   \n",
       "...         ...        ...         ...              ...   \n",
       "25872  0.010982   0.000000         600            False   \n",
       "25874  0.020731   0.006923        1257            False   \n",
       "25875  0.020731   0.002575        1262            False   \n",
       "25873  0.020731   0.349754         573            False   \n",
       "25876  0.020731   0.389690         550            False   \n",
       "\n",
       "       any_annotated_in_cluster  any_gene_expressed  protein_coding  \n",
       "0                          True                True            True  \n",
       "1                          True                True            True  \n",
       "2                          True                True            True  \n",
       "3                          True                True            True  \n",
       "4                         False                True             NaN  \n",
       "...                         ...                 ...             ...  \n",
       "25872                      True                True            True  \n",
       "25874                      True                True            True  \n",
       "25875                      True                True            True  \n",
       "25873                      True                True            True  \n",
       "25876                      True                True            True  \n",
       "\n",
       "[25877 rows x 23 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bf809277",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood-050623/filtered_junctions.gor',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4d79aff4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Startx</th>\n",
       "      <th>End</th>\n",
       "      <th>num_Individs</th>\n",
       "      <th>sum_Fcount_passed</th>\n",
       "      <th>sum_Max_min_overhang_passed</th>\n",
       "      <th>sum_Left_entropy_passed</th>\n",
       "      <th>sum_Right_entropy_passed</th>\n",
       "      <th>max_num_Tests_passed</th>\n",
       "      <th>sum_F_OH</th>\n",
       "      <th>sum_F_OH_LE</th>\n",
       "      <th>sum_F_OH_LE_RE</th>\n",
       "      <th>filtered__End</th>\n",
       "      <th>filtered__Strand</th>\n",
       "      <th>filtered__Gene_name</th>\n",
       "      <th>filtered__max_Fcount</th>\n",
       "      <th>filtered__med_Prop_multy_mapped</th>\n",
       "      <th>filtered__expr_cutoff</th>\n",
       "      <th>filtered__max_Fcount_passed</th>\n",
       "      <th>filtered__med_Prop_multy_mapped_passed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943058</td>\n",
       "      <td>943252</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>943252</td>\n",
       "      <td>+</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011258</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943907</td>\n",
       "      <td>139</td>\n",
       "      <td>3</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>943907</td>\n",
       "      <td>+</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011258</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom  Startx     End  num_Individs  sum_Fcount_passed  \\\n",
       "47  chr1  943058  943252            37                  2   \n",
       "48  chr1  943808  943907           139                  3   \n",
       "\n",
       "    sum_Max_min_overhang_passed  sum_Left_entropy_passed  \\\n",
       "47                           31                        1   \n",
       "48                           56                        1   \n",
       "\n",
       "    sum_Right_entropy_passed  max_num_Tests_passed  sum_F_OH  sum_F_OH_LE  \\\n",
       "47                         1                     4         2            1   \n",
       "48                         1                     4         3            1   \n",
       "\n",
       "    sum_F_OH_LE_RE  filtered__End filtered__Strand filtered__Gene_name  \\\n",
       "47               1         943252                +              SAMD11   \n",
       "48               1         943907                +              SAMD11   \n",
       "\n",
       "    filtered__max_Fcount  filtered__med_Prop_multy_mapped  \\\n",
       "47                     4                              0.0   \n",
       "48                     6                              0.0   \n",
       "\n",
       "    filtered__expr_cutoff  filtered__max_Fcount_passed  \\\n",
       "47               0.011258                            1   \n",
       "48               0.011258                            1   \n",
       "\n",
       "    filtered__med_Prop_multy_mapped_passed  \n",
       "47                                       1  \n",
       "48                                       1  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[df2['filtered__Gene_name']=='SAMD11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b695f592",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCombinedLeafCutterDF(chrm,chrmToLeafcutterFiles):\n",
    "    files = chrmToLeafcutterFiles[chrm]\n",
    "    for i in range(len(files)):\n",
    "        if i == 0:\n",
    "            df = pd.read_csv(files[i],sep='\\t')\n",
    "        else:\n",
    "            df = pd.concat([df,pd.read_csv(files[i],sep='\\t')],axis=0)\n",
    "            df = df.drop_duplicates('splice_event_id')\n",
    "    #df = df[['Chrom','Start','End','Gene_id','ClusterID']]\n",
    "    #df = pd.concat([rna_splice_junctions[rna_splice_junctions['Chrom']==chrm],df],axis=0)\n",
    "    return df.sort_values('Start')\n",
    "\n",
    "leafcutterFiles = glob('/odinn/tmp/bjarnih/RNA/leafCutter/GTEx/*/meta_results/clusters_*_summary.tab')\n",
    "chrmToLeafcutterFiles = defaultdict(list)\n",
    "for file in leafcutterFiles:\n",
    "    chrm = file.split('/')[-1].split('_')[1]\n",
    "    chrmToLeafcutterFiles[chrm].append(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e1a152ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = getCombinedLeafCutterDF('chr1',chrmToLeafcutterFiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a910219f",
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_junctions = getJunctions(gtf_gencode,gtf_gencode['ENST00000616016.5'],'-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "24b33d2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[944574, 943698],\n",
       "       [943808, 943253],\n",
       "       [943377, 942559],\n",
       "       [943058, 942410],\n",
       "       [942488, 942136],\n",
       "       [942251, 941144],\n",
       "       [941306, 939275],\n",
       "       [939412, 939040],\n",
       "       [939129, 935772],\n",
       "       [935896, 931039],\n",
       "       [931089, 930155],\n",
       "       [930336, 925922],\n",
       "       [926013, 923923]])"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_junctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1f3d98a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ClusterID</th>\n",
       "      <th>splice_event_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_start_min</th>\n",
       "      <th>Gene_end_max</th>\n",
       "      <th>...</th>\n",
       "      <th>is_annotated</th>\n",
       "      <th>cluster_type</th>\n",
       "      <th>median_psi</th>\n",
       "      <th>max_psi</th>\n",
       "      <th>prop_NA</th>\n",
       "      <th>prop_zero</th>\n",
       "      <th>num_unique</th>\n",
       "      <th>low_variability</th>\n",
       "      <th>any_annotated_in_cluster</th>\n",
       "      <th>any_gene_expressed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943893</td>\n",
       "      <td>clu_45656</td>\n",
       "      <td>chr1:943808:943893:clu_45656</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.002688</td>\n",
       "      <td>0.598383</td>\n",
       "      <td>97</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943893</td>\n",
       "      <td>clu_48001</td>\n",
       "      <td>chr1:943808:943893:clu_48001</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.038229</td>\n",
       "      <td>0.884937</td>\n",
       "      <td>38</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943893</td>\n",
       "      <td>clu_49105</td>\n",
       "      <td>chr1:943808:943893:clu_49105</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.017938</td>\n",
       "      <td>0.063158</td>\n",
       "      <td>0.004219</td>\n",
       "      <td>0.156780</td>\n",
       "      <td>173</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943893</td>\n",
       "      <td>clu_45793</td>\n",
       "      <td>chr1:943808:943893:clu_45793</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.010336</td>\n",
       "      <td>0.757180</td>\n",
       "      <td>64</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943893</td>\n",
       "      <td>clu_34660</td>\n",
       "      <td>chr1:943808:943893:clu_34660</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>24</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943908</td>\n",
       "      <td>clu_45431</td>\n",
       "      <td>chr1:943808:943908:clu_45431</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943908</td>\n",
       "      <td>clu_53580</td>\n",
       "      <td>chr1:943808:943908:clu_53580</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.007519</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>64</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943908</td>\n",
       "      <td>clu_45793</td>\n",
       "      <td>chr1:943808:943908:clu_45793</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.010336</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>64</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943908</td>\n",
       "      <td>clu_41290</td>\n",
       "      <td>chr1:943808:943908:clu_41290</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.053892</td>\n",
       "      <td>0.006329</td>\n",
       "      <td>29</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943916</td>\n",
       "      <td>clu_49105</td>\n",
       "      <td>chr1:943808:943916:clu_49105</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>+</td>\n",
       "      <td>923928.0</td>\n",
       "      <td>944581.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>alt_acceptor</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017857</td>\n",
       "      <td>0.004219</td>\n",
       "      <td>0.885593</td>\n",
       "      <td>28</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chrom   Start     End  ClusterID               splice_event_id Gene_name  \\\n",
       "241  chr1  943808  943893  clu_45656  chr1:943808:943893:clu_45656    SAMD11   \n",
       "257  chr1  943808  943893  clu_48001  chr1:943808:943893:clu_48001    SAMD11   \n",
       "234  chr1  943808  943893  clu_49105  chr1:943808:943893:clu_49105    SAMD11   \n",
       "225  chr1  943808  943893  clu_45793  chr1:943808:943893:clu_45793    SAMD11   \n",
       "90   chr1  943808  943893  clu_34660  chr1:943808:943893:clu_34660    SAMD11   \n",
       "..    ...     ...     ...        ...                           ...       ...   \n",
       "216  chr1  943808  943908  clu_45431  chr1:943808:943908:clu_45431    SAMD11   \n",
       "249  chr1  943808  943908  clu_53580  chr1:943808:943908:clu_53580    SAMD11   \n",
       "226  chr1  943808  943908  clu_45793  chr1:943808:943908:clu_45793    SAMD11   \n",
       "158  chr1  943808  943908  clu_41290  chr1:943808:943908:clu_41290    SAMD11   \n",
       "236  chr1  943808  943916  clu_49105  chr1:943808:943916:clu_49105    SAMD11   \n",
       "\n",
       "             Gene_id Strand  Gene_start_min  Gene_end_max  ...  is_annotated  \\\n",
       "241  ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "257  ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "234  ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "225  ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "90   ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "..               ...    ...             ...           ...  ...           ...   \n",
       "216  ENSG00000187634      +        923928.0      944581.0  ...          True   \n",
       "249  ENSG00000187634      +        923928.0      944581.0  ...          True   \n",
       "226  ENSG00000187634      +        923928.0      944581.0  ...          True   \n",
       "158  ENSG00000187634      +        923928.0      944581.0  ...          True   \n",
       "236  ENSG00000187634      +        923928.0      944581.0  ...         False   \n",
       "\n",
       "     cluster_type  median_psi   max_psi   prop_NA  prop_zero  num_unique  \\\n",
       "241  alt_acceptor    0.000000  0.142857  0.002688   0.598383          97   \n",
       "257  alt_acceptor    0.000000  0.400000  0.038229   0.884937          38   \n",
       "234  alt_acceptor    0.017938  0.063158  0.004219   0.156780         173   \n",
       "225  alt_acceptor    0.000000  0.300000  0.010336   0.757180          64   \n",
       "90   alt_acceptor    0.000000  0.500000  0.007937   0.800000          24   \n",
       "..            ...         ...       ...       ...        ...         ...   \n",
       "216  alt_acceptor    1.000000  1.000000  0.000000   0.000000          75   \n",
       "249  alt_acceptor    1.000000  1.000000  0.007519   0.001894          64   \n",
       "226  alt_acceptor    1.000000  1.000000  0.010336   0.000000          64   \n",
       "158  alt_acceptor    1.000000  1.000000  0.053892   0.006329          29   \n",
       "236  alt_acceptor    0.000000  0.017857  0.004219   0.885593          28   \n",
       "\n",
       "     low_variability  any_annotated_in_cluster  any_gene_expressed  \n",
       "241            False                      True                True  \n",
       "257            False                      True                True  \n",
       "234            False                      True                True  \n",
       "225            False                      True                True  \n",
       "90             False                      True                True  \n",
       "..               ...                       ...                 ...  \n",
       "216            False                      True                True  \n",
       "249            False                      True                True  \n",
       "226            False                      True                True  \n",
       "158            False                      True                True  \n",
       "236            False                      True                True  \n",
       "\n",
       "[65 rows x 22 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3[df3['Start']==943808].sort_values('End')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3c6a4afe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Startx</th>\n",
       "      <th>End</th>\n",
       "      <th>num_Individs</th>\n",
       "      <th>sum_Fcount_passed</th>\n",
       "      <th>sum_Max_min_overhang_passed</th>\n",
       "      <th>sum_Left_entropy_passed</th>\n",
       "      <th>sum_Right_entropy_passed</th>\n",
       "      <th>max_num_Tests_passed</th>\n",
       "      <th>sum_F_OH</th>\n",
       "      <th>sum_F_OH_LE</th>\n",
       "      <th>sum_F_OH_LE_RE</th>\n",
       "      <th>filtered__End</th>\n",
       "      <th>filtered__Strand</th>\n",
       "      <th>filtered__Gene_name</th>\n",
       "      <th>filtered__max_Fcount</th>\n",
       "      <th>filtered__med_Prop_multy_mapped</th>\n",
       "      <th>filtered__expr_cutoff</th>\n",
       "      <th>filtered__max_Fcount_passed</th>\n",
       "      <th>filtered__med_Prop_multy_mapped_passed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943058</td>\n",
       "      <td>943252</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>943252</td>\n",
       "      <td>+</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011258</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943808</td>\n",
       "      <td>943907</td>\n",
       "      <td>139</td>\n",
       "      <td>3</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>943907</td>\n",
       "      <td>+</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011258</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom  Startx     End  num_Individs  sum_Fcount_passed  \\\n",
       "47  chr1  943058  943252            37                  2   \n",
       "48  chr1  943808  943907           139                  3   \n",
       "\n",
       "    sum_Max_min_overhang_passed  sum_Left_entropy_passed  \\\n",
       "47                           31                        1   \n",
       "48                           56                        1   \n",
       "\n",
       "    sum_Right_entropy_passed  max_num_Tests_passed  sum_F_OH  sum_F_OH_LE  \\\n",
       "47                         1                     4         2            1   \n",
       "48                         1                     4         3            1   \n",
       "\n",
       "    sum_F_OH_LE_RE  filtered__End filtered__Strand filtered__Gene_name  \\\n",
       "47               1         943252                +              SAMD11   \n",
       "48               1         943907                +              SAMD11   \n",
       "\n",
       "    filtered__max_Fcount  filtered__med_Prop_multy_mapped  \\\n",
       "47                     4                              0.0   \n",
       "48                     6                              0.0   \n",
       "\n",
       "    filtered__expr_cutoff  filtered__max_Fcount_passed  \\\n",
       "47               0.011258                            1   \n",
       "48               0.011258                            1   \n",
       "\n",
       "    filtered__med_Prop_multy_mapped_passed  \n",
       "47                                       1  \n",
       "48                                       1  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[df2['filtered__Gene_name']=='SAMD11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f12aab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_junctions = getJunctions(gtf_gencode,gtf_gencode['ENST00000327044.7'],'-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a001be26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[944800, 945057],\n",
       "       [945146, 945518],\n",
       "       [945653, 946173],\n",
       "       [946286, 946402],\n",
       "       [946545, 948131],\n",
       "       [948232, 948490],\n",
       "       [948603, 951127],\n",
       "       [951238, 952000],\n",
       "       [952139, 952412],\n",
       "       [952600, 953175],\n",
       "       [953288, 953782],\n",
       "       [953892, 954004],\n",
       "       [954082, 955923],\n",
       "       [956013, 956095],\n",
       "       [956215, 956894],\n",
       "       [957025, 957099],\n",
       "       [957273, 958929],\n",
       "       [959081, 959215]])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_junctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b7f9e7b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([945057, 945042])"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3[df3['Start']==944800]['End'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c25d6d16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Startx</th>\n",
       "      <th>End</th>\n",
       "      <th>num_Individs</th>\n",
       "      <th>sum_Fcount_passed</th>\n",
       "      <th>sum_Max_min_overhang_passed</th>\n",
       "      <th>sum_Left_entropy_passed</th>\n",
       "      <th>sum_Right_entropy_passed</th>\n",
       "      <th>max_num_Tests_passed</th>\n",
       "      <th>sum_F_OH</th>\n",
       "      <th>sum_F_OH_LE</th>\n",
       "      <th>sum_F_OH_LE_RE</th>\n",
       "      <th>filtered__End</th>\n",
       "      <th>filtered__Strand</th>\n",
       "      <th>filtered__Gene_name</th>\n",
       "      <th>filtered__max_Fcount</th>\n",
       "      <th>filtered__med_Prop_multy_mapped</th>\n",
       "      <th>filtered__expr_cutoff</th>\n",
       "      <th>filtered__max_Fcount_passed</th>\n",
       "      <th>filtered__med_Prop_multy_mapped_passed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944800</td>\n",
       "      <td>945041</td>\n",
       "      <td>17831</td>\n",
       "      <td>17564</td>\n",
       "      <td>17823</td>\n",
       "      <td>17257</td>\n",
       "      <td>17290</td>\n",
       "      <td>4</td>\n",
       "      <td>17564</td>\n",
       "      <td>17257</td>\n",
       "      <td>17181</td>\n",
       "      <td>945041</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>457</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944800</td>\n",
       "      <td>945056</td>\n",
       "      <td>17844</td>\n",
       "      <td>17807</td>\n",
       "      <td>17844</td>\n",
       "      <td>17707</td>\n",
       "      <td>17727</td>\n",
       "      <td>4</td>\n",
       "      <td>17807</td>\n",
       "      <td>17707</td>\n",
       "      <td>17693</td>\n",
       "      <td>945056</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>762</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944819</td>\n",
       "      <td>945041</td>\n",
       "      <td>1791</td>\n",
       "      <td>48</td>\n",
       "      <td>1642</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>945041</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944819</td>\n",
       "      <td>945056</td>\n",
       "      <td>3471</td>\n",
       "      <td>114</td>\n",
       "      <td>3336</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>113</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>945056</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr1</td>\n",
       "      <td>945146</td>\n",
       "      <td>945517</td>\n",
       "      <td>17848</td>\n",
       "      <td>17839</td>\n",
       "      <td>17847</td>\n",
       "      <td>17822</td>\n",
       "      <td>17821</td>\n",
       "      <td>4</td>\n",
       "      <td>17839</td>\n",
       "      <td>17822</td>\n",
       "      <td>17817</td>\n",
       "      <td>945517</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1</td>\n",
       "      <td>945422</td>\n",
       "      <td>945517</td>\n",
       "      <td>10114</td>\n",
       "      <td>1481</td>\n",
       "      <td>3380</td>\n",
       "      <td>41</td>\n",
       "      <td>148</td>\n",
       "      <td>4</td>\n",
       "      <td>736</td>\n",
       "      <td>41</td>\n",
       "      <td>31</td>\n",
       "      <td>945517</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr1</td>\n",
       "      <td>945653</td>\n",
       "      <td>946146</td>\n",
       "      <td>16402</td>\n",
       "      <td>9686</td>\n",
       "      <td>16067</td>\n",
       "      <td>7628</td>\n",
       "      <td>7449</td>\n",
       "      <td>4</td>\n",
       "      <td>9678</td>\n",
       "      <td>7628</td>\n",
       "      <td>7109</td>\n",
       "      <td>946146</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr1</td>\n",
       "      <td>945653</td>\n",
       "      <td>946172</td>\n",
       "      <td>17847</td>\n",
       "      <td>17841</td>\n",
       "      <td>17846</td>\n",
       "      <td>17810</td>\n",
       "      <td>17806</td>\n",
       "      <td>4</td>\n",
       "      <td>17841</td>\n",
       "      <td>17810</td>\n",
       "      <td>17800</td>\n",
       "      <td>946172</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1349</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>945653</td>\n",
       "      <td>946401</td>\n",
       "      <td>665</td>\n",
       "      <td>11</td>\n",
       "      <td>554</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>946401</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>40</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946286</td>\n",
       "      <td>946401</td>\n",
       "      <td>17845</td>\n",
       "      <td>17813</td>\n",
       "      <td>17845</td>\n",
       "      <td>17734</td>\n",
       "      <td>17737</td>\n",
       "      <td>4</td>\n",
       "      <td>17813</td>\n",
       "      <td>17734</td>\n",
       "      <td>17718</td>\n",
       "      <td>946401</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>926</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946286</td>\n",
       "      <td>948130</td>\n",
       "      <td>1547</td>\n",
       "      <td>25</td>\n",
       "      <td>1287</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946497</td>\n",
       "      <td>948130</td>\n",
       "      <td>6871</td>\n",
       "      <td>713</td>\n",
       "      <td>2463</td>\n",
       "      <td>74</td>\n",
       "      <td>52</td>\n",
       "      <td>4</td>\n",
       "      <td>427</td>\n",
       "      <td>70</td>\n",
       "      <td>34</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>33</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946545</td>\n",
       "      <td>946756</td>\n",
       "      <td>4430</td>\n",
       "      <td>721</td>\n",
       "      <td>4161</td>\n",
       "      <td>473</td>\n",
       "      <td>485</td>\n",
       "      <td>4</td>\n",
       "      <td>713</td>\n",
       "      <td>473</td>\n",
       "      <td>453</td>\n",
       "      <td>946756</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>37</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946545</td>\n",
       "      <td>948130</td>\n",
       "      <td>17844</td>\n",
       "      <td>17794</td>\n",
       "      <td>17844</td>\n",
       "      <td>17688</td>\n",
       "      <td>17702</td>\n",
       "      <td>4</td>\n",
       "      <td>17794</td>\n",
       "      <td>17688</td>\n",
       "      <td>17673</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946610</td>\n",
       "      <td>948130</td>\n",
       "      <td>115</td>\n",
       "      <td>4</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946839</td>\n",
       "      <td>948130</td>\n",
       "      <td>1774</td>\n",
       "      <td>144</td>\n",
       "      <td>1660</td>\n",
       "      <td>49</td>\n",
       "      <td>46</td>\n",
       "      <td>4</td>\n",
       "      <td>142</td>\n",
       "      <td>49</td>\n",
       "      <td>40</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chr1</td>\n",
       "      <td>946864</td>\n",
       "      <td>948130</td>\n",
       "      <td>485</td>\n",
       "      <td>30</td>\n",
       "      <td>417</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>30</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>948130</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948232</td>\n",
       "      <td>948489</td>\n",
       "      <td>17843</td>\n",
       "      <td>17791</td>\n",
       "      <td>17841</td>\n",
       "      <td>17669</td>\n",
       "      <td>17673</td>\n",
       "      <td>4</td>\n",
       "      <td>17790</td>\n",
       "      <td>17669</td>\n",
       "      <td>17651</td>\n",
       "      <td>948489</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>908</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948591</td>\n",
       "      <td>951126</td>\n",
       "      <td>3934</td>\n",
       "      <td>149</td>\n",
       "      <td>3737</td>\n",
       "      <td>37</td>\n",
       "      <td>34</td>\n",
       "      <td>4</td>\n",
       "      <td>145</td>\n",
       "      <td>37</td>\n",
       "      <td>30</td>\n",
       "      <td>951126</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>21</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948594</td>\n",
       "      <td>951126</td>\n",
       "      <td>8622</td>\n",
       "      <td>1225</td>\n",
       "      <td>8226</td>\n",
       "      <td>659</td>\n",
       "      <td>638</td>\n",
       "      <td>4</td>\n",
       "      <td>1218</td>\n",
       "      <td>659</td>\n",
       "      <td>591</td>\n",
       "      <td>951126</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948603</td>\n",
       "      <td>951126</td>\n",
       "      <td>17847</td>\n",
       "      <td>17815</td>\n",
       "      <td>17847</td>\n",
       "      <td>17725</td>\n",
       "      <td>17734</td>\n",
       "      <td>4</td>\n",
       "      <td>17815</td>\n",
       "      <td>17725</td>\n",
       "      <td>17710</td>\n",
       "      <td>951126</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1055</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr1</td>\n",
       "      <td>951238</td>\n",
       "      <td>951999</td>\n",
       "      <td>17843</td>\n",
       "      <td>17772</td>\n",
       "      <td>17842</td>\n",
       "      <td>17659</td>\n",
       "      <td>17653</td>\n",
       "      <td>4</td>\n",
       "      <td>17771</td>\n",
       "      <td>17659</td>\n",
       "      <td>17630</td>\n",
       "      <td>951999</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>896</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chr1</td>\n",
       "      <td>952139</td>\n",
       "      <td>952411</td>\n",
       "      <td>17846</td>\n",
       "      <td>17747</td>\n",
       "      <td>17844</td>\n",
       "      <td>17596</td>\n",
       "      <td>17598</td>\n",
       "      <td>4</td>\n",
       "      <td>17747</td>\n",
       "      <td>17596</td>\n",
       "      <td>17572</td>\n",
       "      <td>952411</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr1</td>\n",
       "      <td>952535</td>\n",
       "      <td>953174</td>\n",
       "      <td>5959</td>\n",
       "      <td>296</td>\n",
       "      <td>5483</td>\n",
       "      <td>57</td>\n",
       "      <td>65</td>\n",
       "      <td>4</td>\n",
       "      <td>290</td>\n",
       "      <td>57</td>\n",
       "      <td>54</td>\n",
       "      <td>953174</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chr1</td>\n",
       "      <td>952600</td>\n",
       "      <td>953174</td>\n",
       "      <td>17844</td>\n",
       "      <td>17780</td>\n",
       "      <td>17843</td>\n",
       "      <td>17674</td>\n",
       "      <td>17668</td>\n",
       "      <td>4</td>\n",
       "      <td>17779</td>\n",
       "      <td>17674</td>\n",
       "      <td>17641</td>\n",
       "      <td>953174</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>890</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr1</td>\n",
       "      <td>953288</td>\n",
       "      <td>953781</td>\n",
       "      <td>17841</td>\n",
       "      <td>17708</td>\n",
       "      <td>17838</td>\n",
       "      <td>17492</td>\n",
       "      <td>17492</td>\n",
       "      <td>4</td>\n",
       "      <td>17707</td>\n",
       "      <td>17492</td>\n",
       "      <td>17450</td>\n",
       "      <td>953781</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>781</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1</td>\n",
       "      <td>953892</td>\n",
       "      <td>954003</td>\n",
       "      <td>17826</td>\n",
       "      <td>17623</td>\n",
       "      <td>17823</td>\n",
       "      <td>17370</td>\n",
       "      <td>17374</td>\n",
       "      <td>4</td>\n",
       "      <td>17622</td>\n",
       "      <td>17370</td>\n",
       "      <td>17317</td>\n",
       "      <td>954003</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr1</td>\n",
       "      <td>954082</td>\n",
       "      <td>954463</td>\n",
       "      <td>8092</td>\n",
       "      <td>1485</td>\n",
       "      <td>7878</td>\n",
       "      <td>1043</td>\n",
       "      <td>1064</td>\n",
       "      <td>4</td>\n",
       "      <td>1482</td>\n",
       "      <td>1043</td>\n",
       "      <td>993</td>\n",
       "      <td>954463</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr1</td>\n",
       "      <td>954082</td>\n",
       "      <td>955637</td>\n",
       "      <td>2088</td>\n",
       "      <td>28</td>\n",
       "      <td>1641</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>28</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>955637</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr1</td>\n",
       "      <td>954082</td>\n",
       "      <td>955922</td>\n",
       "      <td>17824</td>\n",
       "      <td>17468</td>\n",
       "      <td>17820</td>\n",
       "      <td>17064</td>\n",
       "      <td>17082</td>\n",
       "      <td>4</td>\n",
       "      <td>17468</td>\n",
       "      <td>17064</td>\n",
       "      <td>16988</td>\n",
       "      <td>955922</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>776</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr1</td>\n",
       "      <td>954523</td>\n",
       "      <td>955922</td>\n",
       "      <td>16133</td>\n",
       "      <td>9434</td>\n",
       "      <td>15993</td>\n",
       "      <td>7442</td>\n",
       "      <td>7560</td>\n",
       "      <td>4</td>\n",
       "      <td>9427</td>\n",
       "      <td>7442</td>\n",
       "      <td>7185</td>\n",
       "      <td>955922</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>117</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr1</td>\n",
       "      <td>955477</td>\n",
       "      <td>955922</td>\n",
       "      <td>1385</td>\n",
       "      <td>27</td>\n",
       "      <td>1279</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>955922</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr1</td>\n",
       "      <td>955982</td>\n",
       "      <td>956094</td>\n",
       "      <td>3795</td>\n",
       "      <td>107</td>\n",
       "      <td>3554</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>956094</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr1</td>\n",
       "      <td>956013</td>\n",
       "      <td>956094</td>\n",
       "      <td>17834</td>\n",
       "      <td>17730</td>\n",
       "      <td>17830</td>\n",
       "      <td>17507</td>\n",
       "      <td>17527</td>\n",
       "      <td>4</td>\n",
       "      <td>17730</td>\n",
       "      <td>17507</td>\n",
       "      <td>17470</td>\n",
       "      <td>956094</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>822</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr1</td>\n",
       "      <td>956215</td>\n",
       "      <td>956893</td>\n",
       "      <td>17846</td>\n",
       "      <td>17768</td>\n",
       "      <td>17838</td>\n",
       "      <td>17608</td>\n",
       "      <td>17627</td>\n",
       "      <td>4</td>\n",
       "      <td>17767</td>\n",
       "      <td>17608</td>\n",
       "      <td>17584</td>\n",
       "      <td>956893</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>980</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957025</td>\n",
       "      <td>957098</td>\n",
       "      <td>17835</td>\n",
       "      <td>17651</td>\n",
       "      <td>17832</td>\n",
       "      <td>17383</td>\n",
       "      <td>17413</td>\n",
       "      <td>4</td>\n",
       "      <td>17651</td>\n",
       "      <td>17383</td>\n",
       "      <td>17339</td>\n",
       "      <td>957098</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>961</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957025</td>\n",
       "      <td>958928</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>958928</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957273</td>\n",
       "      <td>958928</td>\n",
       "      <td>17819</td>\n",
       "      <td>17550</td>\n",
       "      <td>17812</td>\n",
       "      <td>17244</td>\n",
       "      <td>17268</td>\n",
       "      <td>4</td>\n",
       "      <td>17550</td>\n",
       "      <td>17244</td>\n",
       "      <td>17200</td>\n",
       "      <td>958928</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>872</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.846592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959081</td>\n",
       "      <td>959178</td>\n",
       "      <td>53</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>959178</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.184011</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959081</td>\n",
       "      <td>959214</td>\n",
       "      <td>17821</td>\n",
       "      <td>17531</td>\n",
       "      <td>17812</td>\n",
       "      <td>17074</td>\n",
       "      <td>16967</td>\n",
       "      <td>4</td>\n",
       "      <td>17531</td>\n",
       "      <td>17074</td>\n",
       "      <td>16901</td>\n",
       "      <td>959214</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.858634</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959136</td>\n",
       "      <td>959214</td>\n",
       "      <td>2363</td>\n",
       "      <td>60</td>\n",
       "      <td>2206</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>57</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>959214</td>\n",
       "      <td>-</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.858634</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom  Startx     End  num_Individs  sum_Fcount_passed  \\\n",
       "49  chr1  944800  945041         17831              17564   \n",
       "50  chr1  944800  945056         17844              17807   \n",
       "51  chr1  944819  945041          1791                 48   \n",
       "52  chr1  944819  945056          3471                114   \n",
       "53  chr1  945146  945517         17848              17839   \n",
       "54  chr1  945422  945517         10114               1481   \n",
       "55  chr1  945653  946146         16402               9686   \n",
       "56  chr1  945653  946172         17847              17841   \n",
       "57  chr1  945653  946401           665                 11   \n",
       "58  chr1  946286  946401         17845              17813   \n",
       "59  chr1  946286  948130          1547                 25   \n",
       "60  chr1  946497  948130          6871                713   \n",
       "61  chr1  946545  946756          4430                721   \n",
       "62  chr1  946545  948130         17844              17794   \n",
       "63  chr1  946610  948130           115                  4   \n",
       "64  chr1  946839  948130          1774                144   \n",
       "65  chr1  946864  948130           485                 30   \n",
       "66  chr1  948232  948489         17843              17791   \n",
       "67  chr1  948591  951126          3934                149   \n",
       "68  chr1  948594  951126          8622               1225   \n",
       "69  chr1  948603  951126         17847              17815   \n",
       "70  chr1  951238  951999         17843              17772   \n",
       "71  chr1  952139  952411         17846              17747   \n",
       "72  chr1  952535  953174          5959                296   \n",
       "73  chr1  952600  953174         17844              17780   \n",
       "74  chr1  953288  953781         17841              17708   \n",
       "75  chr1  953892  954003         17826              17623   \n",
       "76  chr1  954082  954463          8092               1485   \n",
       "77  chr1  954082  955637          2088                 28   \n",
       "78  chr1  954082  955922         17824              17468   \n",
       "79  chr1  954523  955922         16133               9434   \n",
       "80  chr1  955477  955922          1385                 27   \n",
       "81  chr1  955982  956094          3795                107   \n",
       "82  chr1  956013  956094         17834              17730   \n",
       "83  chr1  956215  956893         17846              17768   \n",
       "84  chr1  957025  957098         17835              17651   \n",
       "85  chr1  957025  958928            35                  2   \n",
       "86  chr1  957273  958928         17819              17550   \n",
       "87  chr1  959081  959178            53                  2   \n",
       "88  chr1  959081  959214         17821              17531   \n",
       "89  chr1  959136  959214          2363                 60   \n",
       "\n",
       "    sum_Max_min_overhang_passed  sum_Left_entropy_passed  \\\n",
       "49                        17823                    17257   \n",
       "50                        17844                    17707   \n",
       "51                         1642                        2   \n",
       "52                         3336                       13   \n",
       "53                        17847                    17822   \n",
       "54                         3380                       41   \n",
       "55                        16067                     7628   \n",
       "56                        17846                    17810   \n",
       "57                          554                        4   \n",
       "58                        17845                    17734   \n",
       "59                         1287                        6   \n",
       "60                         2463                       74   \n",
       "61                         4161                      473   \n",
       "62                        17844                    17688   \n",
       "63                          115                        1   \n",
       "64                         1660                       49   \n",
       "65                          417                        9   \n",
       "66                        17841                    17669   \n",
       "67                         3737                       37   \n",
       "68                         8226                      659   \n",
       "69                        17847                    17725   \n",
       "70                        17842                    17659   \n",
       "71                        17844                    17596   \n",
       "72                         5483                       57   \n",
       "73                        17843                    17674   \n",
       "74                        17838                    17492   \n",
       "75                        17823                    17370   \n",
       "76                         7878                     1043   \n",
       "77                         1641                        4   \n",
       "78                        17820                    17064   \n",
       "79                        15993                     7442   \n",
       "80                         1279                        2   \n",
       "81                         3554                       10   \n",
       "82                        17830                    17507   \n",
       "83                        17838                    17608   \n",
       "84                        17832                    17383   \n",
       "85                           35                        1   \n",
       "86                        17812                    17244   \n",
       "87                           53                        1   \n",
       "88                        17812                    17074   \n",
       "89                         2206                        2   \n",
       "\n",
       "    sum_Right_entropy_passed  max_num_Tests_passed  sum_F_OH  sum_F_OH_LE  \\\n",
       "49                     17290                     4     17564        17257   \n",
       "50                     17727                     4     17807        17707   \n",
       "51                         2                     4        45            2   \n",
       "52                        16                     4       113           13   \n",
       "53                     17821                     4     17839        17822   \n",
       "54                       148                     4       736           41   \n",
       "55                      7449                     4      9678         7628   \n",
       "56                     17806                     4     17841        17810   \n",
       "57                         4                     4         9            4   \n",
       "58                     17737                     4     17813        17734   \n",
       "59                         6                     4        25            6   \n",
       "60                        52                     4       427           70   \n",
       "61                       485                     4       713          473   \n",
       "62                     17702                     4     17794        17688   \n",
       "63                         1                     4         4            1   \n",
       "64                        46                     4       142           49   \n",
       "65                        11                     4        30            9   \n",
       "66                     17673                     4     17790        17669   \n",
       "67                        34                     4       145           37   \n",
       "68                       638                     4      1218          659   \n",
       "69                     17734                     4     17815        17725   \n",
       "70                     17653                     4     17771        17659   \n",
       "71                     17598                     4     17747        17596   \n",
       "72                        65                     4       290           57   \n",
       "73                     17668                     4     17779        17674   \n",
       "74                     17492                     4     17707        17492   \n",
       "75                     17374                     4     17622        17370   \n",
       "76                      1064                     4      1482         1043   \n",
       "77                         5                     4        28            4   \n",
       "78                     17082                     4     17468        17064   \n",
       "79                      7560                     4      9427         7442   \n",
       "80                         2                     4        25            2   \n",
       "81                        10                     4       105           10   \n",
       "82                     17527                     4     17730        17507   \n",
       "83                     17627                     4     17767        17608   \n",
       "84                     17413                     4     17651        17383   \n",
       "85                         1                     4         2            1   \n",
       "86                     17268                     4     17550        17244   \n",
       "87                         1                     4         2            1   \n",
       "88                     16967                     4     17531        17074   \n",
       "89                         4                     4        57            2   \n",
       "\n",
       "    sum_F_OH_LE_RE  filtered__End filtered__Strand filtered__Gene_name  \\\n",
       "49           17181         945041                -               NOC2L   \n",
       "50           17693         945056                -               NOC2L   \n",
       "51               2         945041                -               NOC2L   \n",
       "52              11         945056                -               NOC2L   \n",
       "53           17817         945517                -               NOC2L   \n",
       "54              31         945517                -               NOC2L   \n",
       "55            7109         946146                -               NOC2L   \n",
       "56           17800         946172                -               NOC2L   \n",
       "57               4         946401                -               NOC2L   \n",
       "58           17718         946401                -               NOC2L   \n",
       "59               5         948130                -               NOC2L   \n",
       "60              34         948130                -               NOC2L   \n",
       "61             453         946756                -               NOC2L   \n",
       "62           17673         948130                -               NOC2L   \n",
       "63               1         948130                -               NOC2L   \n",
       "64              40         948130                -               NOC2L   \n",
       "65               8         948130                -               NOC2L   \n",
       "66           17651         948489                -               NOC2L   \n",
       "67              30         951126                -               NOC2L   \n",
       "68             591         951126                -               NOC2L   \n",
       "69           17710         951126                -               NOC2L   \n",
       "70           17630         951999                -               NOC2L   \n",
       "71           17572         952411                -               NOC2L   \n",
       "72              54         953174                -               NOC2L   \n",
       "73           17641         953174                -               NOC2L   \n",
       "74           17450         953781                -               NOC2L   \n",
       "75           17317         954003                -               NOC2L   \n",
       "76             993         954463                -               NOC2L   \n",
       "77               3         955637                -               NOC2L   \n",
       "78           16988         955922                -               NOC2L   \n",
       "79            7185         955922                -               NOC2L   \n",
       "80               1         955922                -               NOC2L   \n",
       "81               8         956094                -               NOC2L   \n",
       "82           17470         956094                -               NOC2L   \n",
       "83           17584         956893                -               NOC2L   \n",
       "84           17339         957098                -               NOC2L   \n",
       "85               1         958928                -               NOC2L   \n",
       "86           17200         958928                -               NOC2L   \n",
       "87               1         959178                -               NOC2L   \n",
       "88           16901         959214                -               NOC2L   \n",
       "89               2         959214                -               NOC2L   \n",
       "\n",
       "    filtered__max_Fcount  filtered__med_Prop_multy_mapped  \\\n",
       "49                   457                              0.0   \n",
       "50                   762                              0.0   \n",
       "51                     8                              0.0   \n",
       "52                    13                              0.0   \n",
       "53                  1497                              0.0   \n",
       "54                    30                              0.0   \n",
       "55                    89                              0.0   \n",
       "56                  1349                              0.0   \n",
       "57                    40                              0.0   \n",
       "58                   926                              0.0   \n",
       "59                     9                              0.0   \n",
       "60                    33                              0.0   \n",
       "61                    37                              0.0   \n",
       "62                   871                              0.0   \n",
       "63                     6                              0.0   \n",
       "64                    23                              0.0   \n",
       "65                     8                              0.0   \n",
       "66                   908                              0.0   \n",
       "67                    21                              0.0   \n",
       "68                    18                              0.0   \n",
       "69                  1055                              0.0   \n",
       "70                   896                              0.0   \n",
       "71                   802                              0.0   \n",
       "72                    16                              0.0   \n",
       "73                   890                              0.0   \n",
       "74                   781                              0.0   \n",
       "75                   732                              0.0   \n",
       "76                    30                              0.0   \n",
       "77                     8                              0.0   \n",
       "78                   776                              0.0   \n",
       "79                   117                              0.0   \n",
       "80                    10                              0.0   \n",
       "81                    16                              0.0   \n",
       "82                   822                              0.0   \n",
       "83                   980                              0.0   \n",
       "84                   961                              0.0   \n",
       "85                    30                              0.0   \n",
       "86                   872                              0.0   \n",
       "87                    16                              0.0   \n",
       "88                   860                              0.0   \n",
       "89                    14                              0.0   \n",
       "\n",
       "    filtered__expr_cutoff  filtered__max_Fcount_passed  \\\n",
       "49               1.846592                            1   \n",
       "50               1.846592                            1   \n",
       "51               1.846592                            1   \n",
       "52               1.846592                            1   \n",
       "53               1.846592                            1   \n",
       "54               1.846592                            1   \n",
       "55               1.846592                            1   \n",
       "56               1.846592                            1   \n",
       "57               1.846592                            1   \n",
       "58               1.846592                            1   \n",
       "59               1.846592                            1   \n",
       "60               1.846592                            1   \n",
       "61               1.846592                            1   \n",
       "62               1.846592                            1   \n",
       "63               1.846592                            1   \n",
       "64               1.846592                            1   \n",
       "65               1.846592                            1   \n",
       "66               1.846592                            1   \n",
       "67               1.846592                            1   \n",
       "68               1.846592                            1   \n",
       "69               1.846592                            1   \n",
       "70               1.846592                            1   \n",
       "71               1.846592                            1   \n",
       "72               1.846592                            1   \n",
       "73               1.846592                            1   \n",
       "74               1.846592                            1   \n",
       "75               1.846592                            1   \n",
       "76               1.846592                            1   \n",
       "77               1.846592                            1   \n",
       "78               1.846592                            1   \n",
       "79               1.846592                            1   \n",
       "80               1.846592                            1   \n",
       "81               1.846592                            1   \n",
       "82               1.846592                            1   \n",
       "83               1.846592                            1   \n",
       "84               1.846592                            1   \n",
       "85               1.846592                            1   \n",
       "86               1.846592                            1   \n",
       "87               1.184011                            1   \n",
       "88               0.858634                            1   \n",
       "89               0.858634                            1   \n",
       "\n",
       "    filtered__med_Prop_multy_mapped_passed  \n",
       "49                                       1  \n",
       "50                                       1  \n",
       "51                                       1  \n",
       "52                                       1  \n",
       "53                                       1  \n",
       "54                                       1  \n",
       "55                                       1  \n",
       "56                                       1  \n",
       "57                                       1  \n",
       "58                                       1  \n",
       "59                                       1  \n",
       "60                                       1  \n",
       "61                                       1  \n",
       "62                                       1  \n",
       "63                                       1  \n",
       "64                                       1  \n",
       "65                                       1  \n",
       "66                                       1  \n",
       "67                                       1  \n",
       "68                                       1  \n",
       "69                                       1  \n",
       "70                                       1  \n",
       "71                                       1  \n",
       "72                                       1  \n",
       "73                                       1  \n",
       "74                                       1  \n",
       "75                                       1  \n",
       "76                                       1  \n",
       "77                                       1  \n",
       "78                                       1  \n",
       "79                                       1  \n",
       "80                                       1  \n",
       "81                                       1  \n",
       "82                                       1  \n",
       "83                                       1  \n",
       "84                                       1  \n",
       "85                                       1  \n",
       "86                                       1  \n",
       "87                                       1  \n",
       "88                                       1  \n",
       "89                                       1  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[df2['filtered__Gene_name']=='NOC2L']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bcc0012a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fasta = pyfastx.Fasta('/odinn/tmp/benediktj/Data/SplicePrediction-050422/genome.fa')\n",
    "gtf_gencode = gffutils.FeatureDB('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/gencode.v42.annotation.db')\n",
    "fasta = pyfastx.Fasta('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/GRCh38.p13.genome.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e3714b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood'\n",
    "\n",
    "seqData = {}\n",
    "\n",
    "CHROM_GROUP = ['chr1', 'chr3', 'chr5', 'chr7', 'chr9',\n",
    "'chr11', 'chr13', 'chr15', 'chr17', 'chr19', 'chr21',\n",
    "'chr2', 'chr4', 'chr6', 'chr8', 'chr10', 'chr12',\n",
    "'chr14', 'chr16', 'chr18', 'chr20', 'chr22', 'chrX', 'chrY']\n",
    "\n",
    "for chrom in CHROM_GROUP:\n",
    "        seqData[chrom] = dok_matrix((len(fasta[chrom]), 5), dtype=np.int8)\n",
    "\n",
    "seqData[chrom] = dok_matrix((len(fasta[chrom]), 5), dtype=np.int8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "72c12f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_datapoints(seq, strand, tx_start, tx_end):\n",
    "    # This function first converts the sequence into an integer array, where\n",
    "    # A, C, G, T, Missing are mapped to 1, 2, 3, 4, 5 respectively. If the strand is\n",
    "    # negative, then reverse complementing is done. . It then calls reformat_data and one_hot_encode\n",
    "\n",
    "    seq = seq.upper()\n",
    "    seq = re.sub(r'[^AGTC]', '5',seq)\n",
    "    seq = seq.replace('A', '1').replace('C', '2')\n",
    "    seq = seq.replace('G', '3').replace('T', '4')\n",
    "\n",
    "    tx_start = int(tx_start)\n",
    "    tx_end = int(tx_end) \n",
    "\n",
    "    Y_idx = []\n",
    "    \n",
    "    X0 = np.asarray([int(x) for x in seq])\n",
    "\n",
    "    X = one_hot_encode(X0)\n",
    "\n",
    "    return X\n",
    "\n",
    "def ceil_div(x, y):\n",
    "    return int(ceil(float(x)/y))\n",
    "\n",
    "\n",
    "IN_MAP = np.asarray([[0, 0, 0, 0,0],\n",
    "                     [1, 0, 0, 0,0],\n",
    "                     [0, 1, 0, 0,0],\n",
    "                     [0, 0, 1, 0,0],\n",
    "                     [0, 0, 0, 1,0],\n",
    "                    [0, 0, 0, 0,1]])\n",
    "# One-hot encoding of the inputs: 0 is for padding, and 1, 2, 3, 4 correspond\n",
    "# to A, C, G, T, Missing respectively.\n",
    "\n",
    "OUT_MAP = np.asarray([[1, 0, 0],\n",
    "                      [0, 1, 0],\n",
    "                      [0, 0, 1],\n",
    "                      [0, 0, 0]])\n",
    "\n",
    "def one_hot_encode(Xd):\n",
    "    return IN_MAP[Xd.astype('int8')]\n",
    "\n",
    "def getJunctions(gtf,transcript,strand):\n",
    "    #transcript = gtf[transcript_id.split('.')[0]]\n",
    "    exon_junctions = []\n",
    "    tx_start = int(transcript[3])\n",
    "    tx_end = int(transcript[4])\n",
    "    exons = gtf.children(transcript, featuretype=\"exon\")\n",
    "    for exon in exons:\n",
    "        exon_start = int(exon[3])\n",
    "        exon_end = int(exon[4])\n",
    "        exon_junctions.append((exon_start,exon_end))\n",
    "\n",
    "    intron_junctions = []\n",
    "\n",
    "    if strand=='+':\n",
    "        intron_start = exon_junctions[0][1]\n",
    "        for i,exon_junction in enumerate(exon_junctions[1:]):\n",
    "            intron_end = exon_junction[0]\n",
    "            intron_junctions.append((intron_start,intron_end))\n",
    "            if i+1 != len(exon_junctions[1:]):\n",
    "                intron_start = exon_junction[1]\n",
    "\n",
    "    elif strand=='-':\n",
    "        exon_junctions.reverse()\n",
    "        intron_start = exon_junctions[0][1]\n",
    "        for i,exon_junction in enumerate(exon_junctions[1:]):\n",
    "            intron_end = exon_junction[0]\n",
    "            intron_junctions.append((intron_start,intron_end))\n",
    "            if i+1 != len(exon_junctions[1:]):\n",
    "                intron_start = exon_junction[1]\n",
    "\n",
    "    return np.array(intron_junctions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "73242073",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove ‘/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/annotation_rnasplice-blood.txt’: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!rm /odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood/annotation_rnasplice-blood.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ce92d774",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "243275it [3:02:00, 22.28it/s] \n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'df' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_79625/1591498870.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mintron_junctions\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mchrom\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0mprev_chrom\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 23\u001b[0;31m                 \u001b[0mleaf_cutter_junctions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetCombinedLeafCutterDF\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mchrmToLeafcutterFiles\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     24\u001b[0m                 \u001b[0mgene_ids_containing_junction\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mleaf_cutter_junctions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Gene_id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mjunction\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mintron_junctions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_79625/3823032878.py\u001b[0m in \u001b[0;36mgetCombinedLeafCutterDF\u001b[0;34m(chrm, chrmToLeafcutterFiles)\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiles\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m             \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_duplicates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'splice_event_id'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Start'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'df' referenced before assignment"
     ]
    }
   ],
   "source": [
    "transcripts = gtf_gencode.features_of_type('transcript')\n",
    "gene_to_label = {}\n",
    "save_seq = True\n",
    "prev_chrom = 'chr1'\n",
    "leaf_cutter_junctions = getCombinedLeafCutterDF(prev_chrom,chrmToLeafcutterFiles)\n",
    "gene_ids_containing_junction = leaf_cutter_junctions['Gene_id'].apply(lambda x: x.split(',')).values\n",
    "for transcript in tqdm(transcripts): \n",
    "    chrom,gene_start,gene_end,strand,gene_id,transcript_id,gene_type,gene_name,level = transcript[0],transcript[3],transcript[4],transcript[6],transcript[8]['gene_id'][0],transcript[8]['transcript_id'][0],transcript[8]['gene_type'][0],transcript[8]['gene_name'][0],transcript[8]['level'][0]\n",
    "    \n",
    "    try:\n",
    "        cond1 = 'Ensembl_canonical' in transcript[8]['tag']\n",
    "    except:\n",
    "        continue\n",
    "    cond2 = gene_type=='protein_coding'\n",
    "    cond3 = int(level)<3\n",
    "    if cond1 and cond2 and cond3:\n",
    "        intron_junctions = getJunctions(gtf_gencode,transcript,strand)\n",
    "        junction_starts = defaultdict(int)\n",
    "        junction_ends = defaultdict(int)\n",
    "\n",
    "        if len(intron_junctions>0):\n",
    "            if chrom!=prev_chrom:\n",
    "                leaf_cutter_junctions = getCombinedLeafCutterDF(chrom,chrmToLeafcutterFiles)\n",
    "                gene_ids_containing_junction = leaf_cutter_junctions['Gene_id'].apply(lambda x: x.split(',')).values\n",
    "            for junction in intron_junctions:\n",
    "                junction_starts[junction[0]] = 1\n",
    "                junction_ends[junction[1]] = 1\n",
    "            simple_gene_id = gene_id.split('.')[0]\n",
    "\n",
    "            alt_junctions = leaf_cutter_junctions[[(simple_gene_id in x) for x in gene_ids_containing_junction]]\n",
    "\n",
    "            if alt_junctions.shape[0]>0:\n",
    "                clusters = defaultdict(int)\n",
    "                for i,pos in enumerate(alt_junctions['Start']):\n",
    "                    if junction_starts[pos] == 1:\n",
    "                        clusters[alt_junctions.iloc[i,:]['ClusterID']] = 1\n",
    "                for i,pos in enumerate(alt_junctions['End']):\n",
    "                    if junction_ends[pos] == 1:\n",
    "                        clusters[alt_junctions.iloc[i,:]['ClusterID']] = 1\n",
    "                for cluster in clusters.keys():\n",
    "                    cluster_junctions = alt_junctions[alt_junctions['ClusterID']==cluster][['Start','End']]\n",
    "                    for i_junc in range(cluster_junctions.shape[0]):\n",
    "                        start,end = cluster_junctions.iloc[i_junc,:]['Start'],cluster_junctions.iloc[i_junc,:]['End']\n",
    "                        junction_id = '{}:{}:{}:{}'.format(chrom,start,end,cluster)\n",
    "                        if not discardJunctionDict[junction_id]:\n",
    "                            junction_starts[start] = 1\n",
    "                            junction_ends[end] = 1\n",
    "\n",
    "                junction_starts = {k:v for k,v in junction_starts.items() if v != 0}\n",
    "                junction_ends = {k:v for k,v in junction_ends.items() if v != 0}\n",
    "                gene_to_label[gene_id] = [junction_starts, junction_ends]\n",
    "\n",
    "                if save_seq:\n",
    "                    seq = fasta[chrom][int(gene_start)-1:int(gene_end)]\n",
    "                    seq = seq.seq\n",
    "                    X = create_datapoints(seq, strand, gene_start, gene_end)\n",
    "                    seqData[chrom][int(gene_start)-1:int(gene_end)] = X\n",
    "                    jn_start = list(junction_starts.keys())\n",
    "                    jn_end = list(junction_ends.keys())\n",
    "                    name = '{}\\t{}\\t{}\\t{}'.format(gene_name,gene_id,transcript_id,level)\n",
    "\n",
    "                    if strand=='+':\n",
    "                        with open('{}/annotation_rnasplice-blood.txt'.format(data_dir), 'a') as the_file:\n",
    "                            the_file.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(name,chrom,strand,gene_start,gene_end,','.join([str(x) for x in jn_start]),','.join([str(x) for x in jn_end])))\n",
    "                    if strand=='-':\n",
    "                        with open('{}/annotation_rnasplice-blood.txt'.format(data_dir,), 'a') as the_file:\n",
    "                            the_file.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(name,chrom,strand,gene_start,gene_end,','.join([str(x) for x in jn_end]),','.join([str(x) for x in jn_start])))\n",
    "\n",
    "                    if chrom!=prev_chrom:\n",
    "                        save_npz('{}/sparse_sequence_data/{}.npz'.format(data_dir,prev_chrom), seqData[prev_chrom].tocoo())\n",
    "                        del seqData[prev_chrom]\n",
    "\n",
    "                    prev_chrom = chrom\n",
    "                \n",
    "    #except:\n",
    "     #   pass\n",
    "        #print(gene[2])\n",
    "        #print(gene[8]['transcript_support_level'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0e46ea13",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_npz('{}/sparse_sequence_data/{}.npz'.format(data_dir,prev_chrom), seqData[prev_chrom].tocoo())\n",
    "#del seqData[prev_chrom]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ae7201e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('{}/gene_to_label.pickle'.format(data_dir), 'wb') as handle:\n",
    "    pickle.dump(gene_to_label, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a410274",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
