{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1c3bae6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benediktj/Python3.9.1/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import gmean\n",
    "from tqdm import tqdm\n",
    "import gffutils\n",
    "import os.path\n",
    "#from liftover import ChainFile\n",
    "from collections import defaultdict\n",
    "from math import ceil\n",
    "import pyfastx\n",
    "import re\n",
    "from scipy.sparse.csgraph import connected_components\n",
    "from glob import glob\n",
    "from scipy.sparse import lil_matrix,csr_matrix,coo_matrix,dok_matrix, save_npz\n",
    "import pickle\n",
    "#\n",
    "blacklist = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/ENCFF220FIN.bed',header=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cab6af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "simpileRepeats = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/simpleRepeat.txt',header=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "88746baa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gffutils.create_db('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX/gencode.v42.annotation.gtf', \"/odinn/tmp/benediktj/Data/SplicePrediction-GTEX/gencode.v42.annotation.db\", force=True,disable_infer_genes=True, disable_infer_transcripts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d829cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_gencode = gffutils.FeatureDB('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/gencode.v42.annotation.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e71ed15",
   "metadata": {},
   "outputs": [],
   "source": [
    "leafcutterFiles = glob('/nfs/odinn/users/solvir/GTEx/GTEx_Analysis_v8_sQTL_leafcutter_counts/*_perind_numers.counts.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b5adf3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#junctions = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions.gct', skiprows=2,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e64dad7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findLowCountJunctions(junctions,blacklist,simpileRepeats):\n",
    "    tmp2 = junctions.iloc[:,2:]\n",
    "    transcriptCount = np.sum(tmp2,axis=1)\n",
    "    tmp2.columns = [x.split('-')[1] for x in junctions.columns[2:]]\n",
    "    tmp2 = tmp2.T.groupby(tmp2.columns).sum().T\n",
    "\n",
    "    includeJunction = np.sum(tmp2 > 0,axis=1)>=4\n",
    "\n",
    "    discardJunctionDict = defaultdict(bool)\n",
    "    discardReason = defaultdict(bool)\n",
    "    prev_chrom = 'chr1'\n",
    "    blacklist_chrom = blacklist[blacklist[0]==prev_chrom]\n",
    "    simpileRepeats_chrom = simpileRepeats[simpileRepeats[1]==prev_chrom]\n",
    "    \n",
    "    for i,junction in tqdm(enumerate(junctions.Name.values)):\n",
    "        chrom,start,end = junctions.iloc[i,0].split('_')\n",
    "        start,end = int(start),int(end)\n",
    "        cond1 = includeJunction[i]==False\n",
    "        \n",
    "        if cond1:\n",
    "            geneID = junctions.iloc[i,1].split('.')[0]\n",
    "            discardJunctionDict[junction+'_'+geneID] = True\n",
    "            discardReason[junction+'_'+geneID] = 'LowReadCount'\n",
    "            prev_chrom = chrom\n",
    "            continue\n",
    "            \n",
    "        if prev_chrom != chrom:\n",
    "            blacklist_chrom = blacklist[blacklist[0]==chrom]\n",
    "        tmp = blacklist_chrom[np.logical_not(np.any([blacklist_chrom[1]>=end, blacklist_chrom[2]<=start],0))]\n",
    "        cond2 = False\n",
    "        for i_b in range(tmp.shape[0]):\n",
    "            if (np.isin(start, range(tmp.iloc[i_b,1],tmp.iloc[i_b,2]+1)) or np.isin(end, range(tmp.iloc[i_b,1],tmp.iloc[i_b,2]+1))):\n",
    "                cond2 = True\n",
    "                break\n",
    "                \n",
    "        if cond2:\n",
    "            geneID = junctions.iloc[i,1].split('.')[0]\n",
    "            discardJunctionDict[junction+'_'+geneID] = True\n",
    "            discardReason[junction+'_'+geneID] = 'InBlacklistedRegion'\n",
    "            prev_chrom = chrom\n",
    "            continue\n",
    "                \n",
    "        if prev_chrom != chrom:\n",
    "            simpileRepeats_chrom = simpileRepeats[simpileRepeats[1]==chrom]\n",
    "        \n",
    "        tmp = simpileRepeats_chrom[np.logical_not(np.any([simpileRepeats_chrom[2]>=end,simpileRepeats_chrom[3]<=start],0))]\n",
    "        cond3 = False\n",
    "        for i_b in range(tmp.shape[0]):\n",
    "            if (np.isin(start, range(tmp.iloc[i_b,2],tmp.iloc[i_b,3]+1)) or np.isin(end, range(tmp.iloc[i_b,2],tmp.iloc[i_b,3]+1))):\n",
    "                cond3 = True\n",
    "                break\n",
    "        \n",
    "        if cond3:\n",
    "            geneID = junctions.iloc[i,1].split('.')[0]\n",
    "            discardJunctionDict[junction+'_'+geneID] = True\n",
    "            discardReason[junction+'_'+geneID] = 'InRepeatRegion'\n",
    "            prev_chrom = chrom\n",
    "            continue\n",
    "        \n",
    "        prev_chrom = chrom\n",
    "        \n",
    "    return discardJunctionDict,discardReason"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98d2fb56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#discardJunctionDict,discardReason = findLowCountJunctions(junctions,blacklist,simpileRepeats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4ac6fdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/discardJunctions.pkl', 'rb') as f:\n",
    "    discardJunctionDict = pickle.load(f)\n",
    "\n",
    "with open('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/discardReason.pkl', 'rb') as f:\n",
    "     discardReason = pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe5b84df",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/discardJunctions.pkl', 'wb') as f:\n",
    "#    pickle.dump(discardJunctionDict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f1d37a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/discardReason.pkl', 'wb') as f:\n",
    "#    pickle.dump(discardReason, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a411b6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "leafcutterFiles = glob('/odinn/tmp/bjarnih/RNA/leafCutter/GTEx/*/meta_results/clusters_*_summary.tab')\n",
    "chrmToLeafcutterFiles = defaultdict(list)\n",
    "for file in leafcutterFiles:\n",
    "    chrm = file.split('/')[-1].split('_')[1]\n",
    "    chrmToLeafcutterFiles[chrm].append(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "edc6fbf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCombinedLeafCutterDF(chrm,chrmToLeafcutterFiles):\n",
    "    files = chrmToLeafcutterFiles[chrm]\n",
    "    for i in range(len(files)):\n",
    "        if i == 0:\n",
    "            df = pd.read_csv(files[i],sep='\\t')\n",
    "        else:\n",
    "            df = pd.concat([df,pd.read_csv(files[i],sep='\\t')],axis=0)\n",
    "            df = df.drop_duplicates('splice_event_id')\n",
    "    df = df[['Chrom','Start','End','Gene_id','ClusterID']]\n",
    "    df = pd.concat([rna_splice_junctions[rna_splice_junctions['Chrom']==chrm],df],axis=0)\n",
    "    return df.sort_values('Start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8511b7d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.read_csv('/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood-050623/filtered_junctions.gor', sep='\\t')\n",
    "df2 = df2.rename(columns={'Startx':'Start'})\n",
    "df2['End'] = df2['End']+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "de2846ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df1 = getCombinedLeafCutterDF('chr1',chrmToLeafcutterFiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f3d9a3a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>ClusterID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13401</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENSG00000223972_rna_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13221</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>clu_93414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13483</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>clu_93414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13052</td>\n",
       "      <td>13219</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENSG00000223972_rna_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14829</td>\n",
       "      <td>15796</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>clu_47199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25509</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248930191</td>\n",
       "      <td>248936582</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>clu_58144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51761</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248935432</td>\n",
       "      <td>248936715</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clu_104118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51762</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248935679</td>\n",
       "      <td>248936582</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>clu_104117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20395</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248935679</td>\n",
       "      <td>248936582</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>clu_46049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51763</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248935699</td>\n",
       "      <td>248936715</td>\n",
       "      <td>NaN</td>\n",
       "      <td>clu_104118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994025 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chrom      Start        End          Gene_id                  ClusterID\n",
       "0      chr1      12721      13401  ENSG00000223972  ENSG00000223972_rna_blood\n",
       "0      chr1      12721      13221  ENSG00000223972                  clu_93414\n",
       "1      chr1      12721      13483  ENSG00000223972                  clu_93414\n",
       "1      chr1      13052      13219  ENSG00000223972  ENSG00000223972_rna_blood\n",
       "3      chr1      14829      15796  ENSG00000227232                  clu_47199\n",
       "...     ...        ...        ...              ...                        ...\n",
       "25509  chr1  248930191  248936582  ENSG00000185220                  clu_58144\n",
       "51761  chr1  248935432  248936715              NaN                 clu_104118\n",
       "51762  chr1  248935679  248936582  ENSG00000185220                 clu_104117\n",
       "20395  chr1  248935679  248936582  ENSG00000185220                  clu_46049\n",
       "51763  chr1  248935699  248936715              NaN                 clu_104118\n",
       "\n",
       "[994025 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b9a02e07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>num_Individs</th>\n",
       "      <th>sum_Fcount_passed</th>\n",
       "      <th>sum_Max_min_overhang_passed</th>\n",
       "      <th>sum_Left_entropy_passed</th>\n",
       "      <th>sum_Right_entropy_passed</th>\n",
       "      <th>max_num_Tests_passed</th>\n",
       "      <th>sum_F_OH</th>\n",
       "      <th>sum_F_OH_LE</th>\n",
       "      <th>sum_F_OH_LE_RE</th>\n",
       "      <th>filtered__End</th>\n",
       "      <th>filtered__Strand</th>\n",
       "      <th>filtered__Gene_name</th>\n",
       "      <th>filtered__max_Fcount</th>\n",
       "      <th>filtered__med_Prop_multy_mapped</th>\n",
       "      <th>filtered__expr_cutoff</th>\n",
       "      <th>filtered__max_Fcount_passed</th>\n",
       "      <th>filtered__med_Prop_multy_mapped_passed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13401</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13402</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.210272</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13052</td>\n",
       "      <td>13219</td>\n",
       "      <td>1450</td>\n",
       "      <td>47</td>\n",
       "      <td>1405</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>13220</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.210272</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183014</td>\n",
       "      <td>2920</td>\n",
       "      <td>165</td>\n",
       "      <td>2741</td>\n",
       "      <td>51</td>\n",
       "      <td>50</td>\n",
       "      <td>4</td>\n",
       "      <td>161</td>\n",
       "      <td>51</td>\n",
       "      <td>44</td>\n",
       "      <td>183015</td>\n",
       "      <td>+</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.209060</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183112</td>\n",
       "      <td>9911</td>\n",
       "      <td>2627</td>\n",
       "      <td>9299</td>\n",
       "      <td>1783</td>\n",
       "      <td>1789</td>\n",
       "      <td>4</td>\n",
       "      <td>2620</td>\n",
       "      <td>1783</td>\n",
       "      <td>1701</td>\n",
       "      <td>183113</td>\n",
       "      <td>+</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>62</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.209060</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>183130</td>\n",
       "      <td>15633</td>\n",
       "      <td>9407</td>\n",
       "      <td>15347</td>\n",
       "      <td>7985</td>\n",
       "      <td>7956</td>\n",
       "      <td>4</td>\n",
       "      <td>9403</td>\n",
       "      <td>7985</td>\n",
       "      <td>7741</td>\n",
       "      <td>183131</td>\n",
       "      <td>+</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>233</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.209060</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386633</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21312980</td>\n",
       "      <td>21317614</td>\n",
       "      <td>1635</td>\n",
       "      <td>71</td>\n",
       "      <td>1316</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>67</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>21317615</td>\n",
       "      <td>+</td>\n",
       "      <td>AC010086.1</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386634</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21314394</td>\n",
       "      <td>21317614</td>\n",
       "      <td>1215</td>\n",
       "      <td>38</td>\n",
       "      <td>1139</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>21317615</td>\n",
       "      <td>+</td>\n",
       "      <td>AC010086.1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386635</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21315341</td>\n",
       "      <td>21317614</td>\n",
       "      <td>1693</td>\n",
       "      <td>63</td>\n",
       "      <td>1591</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>63</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>21317615</td>\n",
       "      <td>+</td>\n",
       "      <td>AC010086.1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386636</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21317726</td>\n",
       "      <td>21317826</td>\n",
       "      <td>6635</td>\n",
       "      <td>2501</td>\n",
       "      <td>6435</td>\n",
       "      <td>1714</td>\n",
       "      <td>1707</td>\n",
       "      <td>4</td>\n",
       "      <td>2496</td>\n",
       "      <td>1714</td>\n",
       "      <td>1584</td>\n",
       "      <td>21317827</td>\n",
       "      <td>+</td>\n",
       "      <td>AC010086.1</td>\n",
       "      <td>227</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386637</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21317896</td>\n",
       "      <td>21318318</td>\n",
       "      <td>6596</td>\n",
       "      <td>2383</td>\n",
       "      <td>6461</td>\n",
       "      <td>1653</td>\n",
       "      <td>1646</td>\n",
       "      <td>4</td>\n",
       "      <td>2381</td>\n",
       "      <td>1653</td>\n",
       "      <td>1559</td>\n",
       "      <td>21318319</td>\n",
       "      <td>+</td>\n",
       "      <td>AC010086.1</td>\n",
       "      <td>245</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>386638 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chrom     Start       End  num_Individs  sum_Fcount_passed  \\\n",
       "0       chr1     12721     13401           171                  1   \n",
       "1       chr1     13052     13219          1450                 47   \n",
       "2       chr1    182746    183014          2920                165   \n",
       "3       chr1    182746    183112          9911               2627   \n",
       "4       chr1    182746    183130         15633               9407   \n",
       "...      ...       ...       ...           ...                ...   \n",
       "386633  chrY  21312980  21317614          1635                 71   \n",
       "386634  chrY  21314394  21317614          1215                 38   \n",
       "386635  chrY  21315341  21317614          1693                 63   \n",
       "386636  chrY  21317726  21317826          6635               2501   \n",
       "386637  chrY  21317896  21318318          6596               2383   \n",
       "\n",
       "        sum_Max_min_overhang_passed  sum_Left_entropy_passed  \\\n",
       "0                               171                        1   \n",
       "1                              1405                        2   \n",
       "2                              2741                       51   \n",
       "3                              9299                     1783   \n",
       "4                             15347                     7985   \n",
       "...                             ...                      ...   \n",
       "386633                         1316                        8   \n",
       "386634                         1139                        3   \n",
       "386635                         1591                        4   \n",
       "386636                         6435                     1714   \n",
       "386637                         6461                     1653   \n",
       "\n",
       "        sum_Right_entropy_passed  max_num_Tests_passed  sum_F_OH  sum_F_OH_LE  \\\n",
       "0                              1                     4         1            1   \n",
       "1                              3                     4        46            2   \n",
       "2                             50                     4       161           51   \n",
       "3                           1789                     4      2620         1783   \n",
       "4                           7956                     4      9403         7985   \n",
       "...                          ...                   ...       ...          ...   \n",
       "386633                         9                     4        67            8   \n",
       "386634                         3                     4        36            3   \n",
       "386635                         4                     4        63            4   \n",
       "386636                      1707                     4      2496         1714   \n",
       "386637                      1646                     4      2381         1653   \n",
       "\n",
       "        sum_F_OH_LE_RE  filtered__End filtered__Strand filtered__Gene_name  \\\n",
       "0                    1          13402                +             DDX11L1   \n",
       "1                    2          13220                +             DDX11L1   \n",
       "2                   44         183015                +          FO538757.2   \n",
       "3                 1701         183113                +          FO538757.2   \n",
       "4                 7741         183131                +          FO538757.2   \n",
       "...                ...            ...              ...                 ...   \n",
       "386633               7       21317615                +          AC010086.1   \n",
       "386634               3       21317615                +          AC010086.1   \n",
       "386635               2       21317615                +          AC010086.1   \n",
       "386636            1584       21317827                +          AC010086.1   \n",
       "386637            1559       21318319                +          AC010086.1   \n",
       "\n",
       "        filtered__max_Fcount  filtered__med_Prop_multy_mapped  \\\n",
       "0                          5                         0.000000   \n",
       "1                         11                         0.000000   \n",
       "2                         16                         0.000000   \n",
       "3                         62                         0.000000   \n",
       "4                        233                         0.076923   \n",
       "...                      ...                              ...   \n",
       "386633                    21                         0.000000   \n",
       "386634                    11                         0.000000   \n",
       "386635                    13                         0.000000   \n",
       "386636                   227                         0.000000   \n",
       "386637                   245                         0.000000   \n",
       "\n",
       "        filtered__expr_cutoff  filtered__max_Fcount_passed  \\\n",
       "0                    0.210272                            1   \n",
       "1                    0.210272                            1   \n",
       "2                    0.209060                            1   \n",
       "3                    0.209060                            1   \n",
       "4                    0.209060                            1   \n",
       "...                       ...                          ...   \n",
       "386633               0.000000                            1   \n",
       "386634               0.000000                            1   \n",
       "386635               0.000000                            1   \n",
       "386636               0.000000                            1   \n",
       "386637               0.000000                            1   \n",
       "\n",
       "        filtered__med_Prop_multy_mapped_passed  \n",
       "0                                            1  \n",
       "1                                            1  \n",
       "2                                            1  \n",
       "3                                            1  \n",
       "4                                            1  \n",
       "...                                        ...  \n",
       "386633                                       1  \n",
       "386634                                       1  \n",
       "386635                                       1  \n",
       "386636                                       1  \n",
       "386637                                       1  \n",
       "\n",
       "[386638 rows x 20 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bcc0012a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = pyfastx.Fasta('/odinn/tmp/benediktj/Data/SplicePrediction-GTEX-V8/GRCh38.p13.genome.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e3714b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/odinn/tmp/benediktj/Data/SplicePrediction-rnasplice-blood-050623/'\n",
    "\n",
    "seqData = {}\n",
    "\n",
    "CHROM_GROUP = ['chr1', 'chr3', 'chr5', 'chr7', 'chr9',\n",
    "'chr11', 'chr13', 'chr15', 'chr17', 'chr19', 'chr21',\n",
    "'chr2', 'chr4', 'chr6', 'chr8', 'chr10', 'chr12',\n",
    "'chr14', 'chr16', 'chr18', 'chr20', 'chr22', 'chrX', 'chrY']\n",
    "\n",
    "for chrom in CHROM_GROUP:\n",
    "        seqData[chrom] = dok_matrix((len(fasta[chrom]), 5), dtype=np.int8)\n",
    "\n",
    "seqData[chrom] = dok_matrix((len(fasta[chrom]), 5), dtype=np.int8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "10e926f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "62696it [01:24, 737.98it/s] \n"
     ]
    }
   ],
   "source": [
    "genes = gtf_gencode.features_of_type('gene')\n",
    "gene_name_to_id = defaultdict(lambda: None)\n",
    "for gene in tqdm(genes):\n",
    "    #print(gene['gene_id'],gene['gene_name'])\n",
    "    gene_name_to_id[gene['gene_name'][0]] = gene['gene_id'][0].split('.')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4269fc13",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2['Gene_id'] = df2['filtered__Gene_name'].apply(lambda x: gene_name_to_id[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f90ec7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_splice_junctions = df2[df2['Gene_id'].apply(lambda x: x is not None)][['Chrom','Start','End','Gene_id']]\n",
    "rna_splice_junctions['ClusterID'] = rna_splice_junctions['Gene_id']+'_rna_blood'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "72c12f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_datapoints(seq, strand, tx_start, tx_end):\n",
    "    # This function first converts the sequence into an integer array, where\n",
    "    # A, C, G, T, Missing are mapped to 1, 2, 3, 4, 5 respectively. If the strand is\n",
    "    # negative, then reverse complementing is done. . It then calls reformat_data and one_hot_encode\n",
    "\n",
    "    seq = seq.upper()\n",
    "    seq = re.sub(r'[^AGTC]', '5',seq)\n",
    "    seq = seq.replace('A', '1').replace('C', '2')\n",
    "    seq = seq.replace('G', '3').replace('T', '4')\n",
    "\n",
    "    tx_start = int(tx_start)\n",
    "    tx_end = int(tx_end) \n",
    "\n",
    "    Y_idx = []\n",
    "    \n",
    "    X0 = np.asarray([int(x) for x in seq])\n",
    "\n",
    "    X = one_hot_encode(X0)\n",
    "\n",
    "    return X\n",
    "\n",
    "def ceil_div(x, y):\n",
    "    return int(ceil(float(x)/y))\n",
    "\n",
    "\n",
    "IN_MAP = np.asarray([[0, 0, 0, 0,0],\n",
    "                     [1, 0, 0, 0,0],\n",
    "                     [0, 1, 0, 0,0],\n",
    "                     [0, 0, 1, 0,0],\n",
    "                     [0, 0, 0, 1,0],\n",
    "                    [0, 0, 0, 0,1]])\n",
    "# One-hot encoding of the inputs: 0 is for padding, and 1, 2, 3, 4 correspond\n",
    "# to A, C, G, T, Missing respectively.\n",
    "\n",
    "OUT_MAP = np.asarray([[1, 0, 0],\n",
    "                      [0, 1, 0],\n",
    "                      [0, 0, 1],\n",
    "                      [0, 0, 0]])\n",
    "\n",
    "def one_hot_encode(Xd):\n",
    "    return IN_MAP[Xd.astype('int8')]\n",
    "\n",
    "def getJunctions(gtf,transcript,strand):\n",
    "    #transcript = gtf[transcript_id.split('.')[0]]\n",
    "    exon_junctions = []\n",
    "    tx_start = int(transcript[3])\n",
    "    tx_end = int(transcript[4])\n",
    "    exons = gtf.children(transcript, featuretype=\"exon\")\n",
    "    for exon in exons:\n",
    "        exon_start = int(exon[3])\n",
    "        exon_end = int(exon[4])\n",
    "        exon_junctions.append((exon_start,exon_end))\n",
    "\n",
    "    intron_junctions = []\n",
    "\n",
    "    if strand=='+':\n",
    "        intron_start = exon_junctions[0][1]\n",
    "        for i,exon_junction in enumerate(exon_junctions[1:]):\n",
    "            intron_end = exon_junction[0]\n",
    "            intron_junctions.append((intron_start,intron_end))\n",
    "            if i+1 != len(exon_junctions[1:]):\n",
    "                intron_start = exon_junction[1]\n",
    "\n",
    "    elif strand=='-':\n",
    "        exon_junctions.reverse()\n",
    "        intron_start = exon_junctions[0][1]\n",
    "        for i,exon_junction in enumerate(exon_junctions[1:]):\n",
    "            intron_end = exon_junction[0]\n",
    "            intron_junctions.append((intron_start,intron_end))\n",
    "            if i+1 != len(exon_junctions[1:]):\n",
    "                intron_start = exon_junction[1]\n",
    "\n",
    "    return np.array(intron_junctions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ce92d774",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "252416it [5:30:35, 12.73it/s]  \n"
     ]
    }
   ],
   "source": [
    "transcripts = gtf_gencode.features_of_type('transcript')\n",
    "gene_to_label = {}\n",
    "save_seq = True\n",
    "prev_chrom = 'chr1'\n",
    "leaf_cutter_junctions = getCombinedLeafCutterDF(prev_chrom,chrmToLeafcutterFiles)\n",
    "for transcript in tqdm(transcripts): \n",
    "    chrom,gene_start,gene_end,strand,gene_id,transcript_id,gene_type,gene_name,level = transcript[0],transcript[3],transcript[4],transcript[6],transcript[8]['gene_id'][0],transcript[8]['transcript_id'][0],transcript[8]['gene_type'][0],transcript[8]['gene_name'][0],transcript[8]['level'][0]\n",
    "    \n",
    "    try:\n",
    "        cond1 = 'Ensembl_canonical' in transcript[8]['tag']\n",
    "        cond2 = gene_type=='protein_coding'\n",
    "        cond3 = int(level)<3\n",
    "        if cond1 and cond2 and cond3:\n",
    "            intron_junctions = getJunctions(gtf_gencode,transcript,strand)\n",
    "            junction_starts = defaultdict(int)\n",
    "            junction_ends = defaultdict(int)\n",
    "            \n",
    "           \n",
    "            if len(intron_junctions>0):\n",
    "                if chrom!=prev_chrom:\n",
    "                    leaf_cutter_junctions = getCombinedLeafCutterDF(chrom,chrmToLeafcutterFiles)\n",
    "                for junction in intron_junctions:\n",
    "                    junction_starts[junction[0]] = 1\n",
    "                    junction_ends[junction[1]] = 1\n",
    "                simple_gene_id = gene_id.split('.')[0]\n",
    "                alt_junctions = leaf_cutter_junctions[leaf_cutter_junctions['Gene_id']==simple_gene_id]\n",
    "                if alt_junctions.shape[0]>0:\n",
    "                    clusters = defaultdict(int)\n",
    "                    for i,pos in enumerate(alt_junctions['Start']):\n",
    "                        if junction_starts[pos] == 1:\n",
    "                            clusters[alt_junctions.iloc[i,:]['ClusterID']] = 1\n",
    "                    for i,pos in enumerate(alt_junctions['End']):\n",
    "                        if junction_ends[pos] == 1:\n",
    "                            clusters[alt_junctions.iloc[i,:]['ClusterID']] = 1\n",
    "                    for cluster in clusters.keys():\n",
    "                        cluster_junctions = alt_junctions[alt_junctions['ClusterID']==cluster][['Start','End']]\n",
    "                        for i_junc in range(cluster_junctions.shape[0]):\n",
    "                            start,end = cluster_junctions.iloc[i_junc,:]['Start'],cluster_junctions.iloc[i_junc,:]['End']\n",
    "                            junction_id = '{}_{}_{}_{}'.format(chrom,start,end,simple_gene_id)\n",
    "                            if not discardJunctionDict[junction_id]:\n",
    "                                junction_starts[start] = 1\n",
    "                                junction_ends[end] = 1\n",
    "                    \n",
    "                    junction_starts = {k:v for k,v in junction_starts.items() if v != 0}\n",
    "                    junction_ends = {k:v for k,v in junction_ends.items() if v != 0}\n",
    "                    gene_to_label[gene_id] = [junction_starts, junction_ends]\n",
    "            \n",
    "                    if save_seq:\n",
    "                        seq = fasta[chrom][int(gene_start)-1:int(gene_end)]\n",
    "                        seq = seq.seq\n",
    "                        X = create_datapoints(seq, strand, gene_start, gene_end)\n",
    "                        seqData[chrom][int(gene_start)-1:int(gene_end)] = X\n",
    "                        jn_start = list(junction_starts.keys())\n",
    "                        jn_end = list(junction_ends.keys())\n",
    "                        name = '{}\\t{}\\t{}\\t{}'.format(gene_name,gene_id,transcript_id,level)\n",
    "\n",
    "                        if strand=='+':\n",
    "                            with open('{}/annotation_GTEX_v8.txt'.format(data_dir), 'a') as the_file:\n",
    "                                the_file.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(name,chrom,strand,gene_start,gene_end,','.join([str(x) for x in jn_start]),','.join([str(x) for x in jn_end])))\n",
    "                        if strand=='-':\n",
    "                            with open('{}/annotation_GTEX_v8.txt'.format(data_dir,), 'a') as the_file:\n",
    "                                the_file.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(name,chrom,strand,gene_start,gene_end,','.join([str(x) for x in jn_end]),','.join([str(x) for x in jn_start])))\n",
    "\n",
    "                        if chrom!=prev_chrom:\n",
    "                            save_npz('{}/sparse_sequence_data/{}.npz'.format(data_dir,prev_chrom), seqData[prev_chrom].tocoo())\n",
    "                            del seqData[prev_chrom]\n",
    "\n",
    "                        prev_chrom = chrom\n",
    "                \n",
    "    except:\n",
    "        pass\n",
    "        #print(gene[2])\n",
    "        #print(gene[8]['transcript_support_level'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0e46ea13",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_npz('{}/sparse_sequence_data/{}.npz'.format(data_dir,prev_chrom), seqData[prev_chrom].tocoo())\n",
    "#del seqData[prev_chrom]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ae7201e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('{}/gene_to_label.pickle'.format(data_dir), 'wb') as handle:\n",
    "    pickle.dump(gene_to_label, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
